{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Atlas Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alejandro has developed an atlas of gene expression across the mouse life span, and this notebook walks through exploratory analysis on this dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Table of Contents\n",
    "\n",
    "- [Introduction](#intro)\n",
    "- [Data Normalization](#norm)\n",
    "- [Scaling of Counts](#scaling)\n",
    "- [Plots](#plots)\n",
    "    - [BMPR](#bmpr)\n",
    "    - [WntR](#wntr)\n",
    "    - [Notch](#notch)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Introduction <a name=\"intro\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This atlas samples a large diversity of mouse cell states, from  early gastrulation (E5.5) to late adulthood (30 months). \n",
    "\n",
    "The individual datasets in the atlas are shown below:\n",
    "\n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;}\n",
    ".tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;\n",
    "  overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;\n",
    "  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n",
    ".tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "<thead>\n",
    "  <tr>\n",
    "    <th class=\"tg-0pky\"><span style=\"font-weight:bold\">Dataset</span></th>\n",
    "    <th class=\"tg-0pky\"><span style=\"font-weight:bold\">Time Points</span></th>\n",
    "    <th class=\"tg-0pky\"><span style=\"font-weight:bold\">Author</span></th>\n",
    "    <th class=\"tg-0pky\"><span style=\"font-weight:bold\">Cells</span></th>\n",
    "    <th class=\"tg-0pky\"><span style=\"font-weight:bold\">Mice Sampled</span></th>\n",
    "    <th class=\"tg-0pky\"><span style=\"font-weight:bold\">Library Type</span></th>\n",
    "  </tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "  <tr>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-weight:bold;font-style:normal;text-decoration:none\">Forelimb</span><span style=\"font-weight:400;font-style:normal;text-decoration:none\"> atlas (</span><span style=\"font-style:normal\">The changing mouse embryo transcriptome at whole tissue and single-cell resolution)</span></td>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-weight:400;font-style:normal\">E10.5, E11.0, E11.5, E12.0, E13.0, E13.5, E14.0, E15.0</span></td>\n",
    "    <td class=\"tg-0pky\">He et al. (2020)</td>\n",
    "    <td class=\"tg-0pky\">90,637</td>\n",
    "    <td class=\"tg-0pky\">Pair of forelimbs per time point</td>\n",
    "    <td class=\"tg-0pky\">10X</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-style:normal\">A single-cell molecular map of </span><span style=\"font-weight:bold;font-style:normal\">mouse gastrulation and early organogenesis</span></td>\n",
    "    <td class=\"tg-0pky\">E6.5, E6.75, E7.0, E7.25, E7.5, E7.75, E8.0, E8.25, E8.5</td>\n",
    "    <td class=\"tg-0pky\">Pijuan-Sala et al. (2019)</td>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-weight:400;font-style:normal\">116,312</span></td>\n",
    "    <td class=\"tg-0pky\">411 mouse embryos total</td>\n",
    "    <td class=\"tg-0pky\">10X</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-style:normal\">The emergent landscape of the mouse gut </span><span style=\"font-weight:bold;font-style:normal\">endoderm</span><span style=\"font-style:normal\"> at single-cell resolution</span></td>\n",
    "    <td class=\"tg-0pky\">E5.5</td>\n",
    "    <td class=\"tg-0pky\">Nowotschin et al. (2019)</td>\n",
    "    <td class=\"tg-0pky\"></td>\n",
    "    <td class=\"tg-0pky\">-</td>\n",
    "    <td class=\"tg-0pky\">10X</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-style:normal\">Single-cell RNA-seq analysis unveils a prevalent epithelial/mesenchymal hybrid state during mouse </span><span style=\"font-weight:bold;font-style:normal\">organogenesis</span></td>\n",
    "    <td class=\"tg-0pky\">E9.5-E11.5</td>\n",
    "    <td class=\"tg-0pky\">Dong et al. (2018)</td>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-weight:400;font-style:normal\">1916</span></td>\n",
    "    <td class=\"tg-0pky\">7 total embryos</td>\n",
    "    <td class=\"tg-0pky\">Smart-seq2</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0lax\"><span style=\"font-style:normal\">Epigenetic regulator function through </span><span style=\"font-weight:700;font-style:normal\">mouse gastrulation</span></td>\n",
    "    <td class=\"tg-0lax\">E6.5, E7.0, E7.5, E8.0, E8.5</td>\n",
    "    <td class=\"tg-0lax\">Grosswendt et al. (2020)</td>\n",
    "    <td class=\"tg-0lax\">88,779</td>\n",
    "    <td class=\"tg-0lax\">50 embryos</td>\n",
    "    <td class=\"tg-0lax\">10X</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0pky\"><span style=\"font-weight:bold\">Tabula Muris and Tabula Muris Senis</span></td>\n",
    "    <td class=\"tg-0pky\">1mo, 3mo, 18mo, 21mo, 24mo, 30mo</td>\n",
    "    <td class=\"tg-0pky\">Tabula Muris Consortium (2018, 2020)</td>\n",
    "    <td class=\"tg-0pky\">450,000+</td>\n",
    "    <td class=\"tg-0pky\">-</td>\n",
    "    <td class=\"tg-0pky\">10X, Smart-seq2</td>\n",
    "  </tr>\n",
    "</tbody>\n",
    "</table>\n",
    "\n",
    "These data were merged by first pre-processing each dataset individually using a standard scRNA-seq pipeline (quality control, filtering, normalization) and then applying unsupervising learning algorithms to cluster cells. Given the excessive number of zeroes, or \"drop-outs,\" in scRNA-seq data, instead of using counts from single-cells in our final atlas, we generated a \"pseudo-bulk\" sample per cluster that corresponds to average gene expression of its members. This generated **1206 unique pseudo-bulk samples from 6 different datasets**. \n",
    "\n",
    "Since we filter out lowly-expressed genes in the scRNA-seq pre-processing step, this atlas does not contain the entire transcriptome. Moreover, since we filter genes separately for each dataset, the atlas only contains the interesction of genes that remain across the pseudo-bulk samples, to **capture a total of 11,675 genes** out of the ~22k in the mouse transcriptome. \n",
    "\n",
    "Our merged data consists of 1206 pseudo-bulk samples, from a diverse range of cell types and stages. We can use this dataset to explore gene expression in various pathways, and how they vary across tissues and time points. This tutorial walks through some additional processing steps and possible analyses for this data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.1.1.min.js\": \"kLr4fYcqcSpbuI95brIH3vnnYCquzzSxHPU6XGQCIkQRGJwhg0StNbj1eegrHs12\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.1.1.min.js\": \"xIGPmVtaOm+z0BqfSOMn4lOR6ciex448GIKG4eE61LsAvmGj48XcMQZtKcE/UXZe\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.1.1.min.js\": \"Dc9u1wF/0zApGIWoBbH77iWEHtdmkuYWG839Uzmv8y8yBLXebjO9ZnERsde5Ln/P\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.1.1.min.js\": \"cT9JaBz7GiRXdENrJLZNSC6eMNF3nh3fa5fTF51Svp+ukxPdwcU5kGXGPBgDCa2j\"};\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      if (url in hashes) {\n",
       "        element.crossOrigin = \"anonymous\";\n",
       "        element.integrity = \"sha384-\" + hashes[url];\n",
       "      }\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.1.1.min.js\"];\n",
       "  var css_urls = [];\n",
       "  \n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (var i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.1.1.min.js\": \"kLr4fYcqcSpbuI95brIH3vnnYCquzzSxHPU6XGQCIkQRGJwhg0StNbj1eegrHs12\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.1.1.min.js\": \"xIGPmVtaOm+z0BqfSOMn4lOR6ciex448GIKG4eE61LsAvmGj48XcMQZtKcE/UXZe\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.1.1.min.js\": \"Dc9u1wF/0zApGIWoBbH77iWEHtdmkuYWG839Uzmv8y8yBLXebjO9ZnERsde5Ln/P\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.1.1.min.js\": \"cT9JaBz7GiRXdENrJLZNSC6eMNF3nh3fa5fTF51Svp+ukxPdwcU5kGXGPBgDCa2j\"};\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      if (url in hashes) {\n        element.crossOrigin = \"anonymous\";\n        element.integrity = \"sha384-\" + hashes[url];\n      }\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.1.1.min.js\"];\n  var css_urls = [];\n  \n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (var i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "import itertools\n",
    "\n",
    "from bokeh.plotting import figure, ColumnDataSource, output_notebook, show, output_file, save\n",
    "from bokeh.models import HoverTool, WheelZoomTool, PanTool, BoxZoomTool, ResetTool, TapTool, SaveTool\n",
    "\n",
    "output_notebook()\n",
    "\n",
    "# For normalization of gene counts\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "import logging\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "\n",
    "# Ignore R warning messages\n",
    "# Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error in library(corrplot) : there is no package called ‘corrplot’\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(corrplot)\n",
    "library(ppcor)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our data files consist of a csv with pseudo-bulk counts from our datasets, a csv with metadata annotations for each pseudo-bulk sample, and a reference list of pathways of interest and their components. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set directories for the data, results, and figures. SPECIFY AS APPROPRIATE\n",
    "datadir = \"./alejo_metadata/data/\"\n",
    "figdir = \"./alejo_metadata/figures/own_norm/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary files\n",
    "# Load the atlas dataset of counts\n",
    "counts = pd.read_csv(datadir + \"fig_version_countsSeurat_Apr22.csv\", index_col = [0]).T\n",
    "\n",
    "# Import metadata csv\n",
    "meta = pd.read_csv(datadir + \"meta_data_Aug2021.csv\", index_col = [0])\n",
    "\n",
    "# Import reference pathway list csv\n",
    "pathways = pd.read_csv(datadir + \"pathway_list.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1_E6.5_8.5_Chan      8459.581496\n",
       "2_E6.5_8.5_Chan      8426.250002\n",
       "3_E6.5_8.5_Chan      8149.569975\n",
       "4_E6.5_8.5_Chan      8545.450894\n",
       "5_E6.5_8.5_Chan      8670.757169\n",
       "                       ...      \n",
       "1202_FACS 3m       215569.839162\n",
       "1203_FACS 3m       280121.144893\n",
       "1204_FACS 3m        85336.773986\n",
       "1205_FACS 3m       107320.202826\n",
       "1206_FACS 3m        88837.805000\n",
       "Length: 1206, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.sum(axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abi2</th>\n",
       "      <th>Acadl</th>\n",
       "      <th>Actr1b</th>\n",
       "      <th>Adam23</th>\n",
       "      <th>Adhfe1</th>\n",
       "      <th>Aff3</th>\n",
       "      <th>Als2</th>\n",
       "      <th>Ankrd39</th>\n",
       "      <th>Ankrd44</th>\n",
       "      <th>Arfgef1</th>\n",
       "      <th>...</th>\n",
       "      <th>Vldlr</th>\n",
       "      <th>Vps13a</th>\n",
       "      <th>Vti1a</th>\n",
       "      <th>Vwa2</th>\n",
       "      <th>Xpnpep1</th>\n",
       "      <th>Zdhhc16</th>\n",
       "      <th>Zdhhc6</th>\n",
       "      <th>Zfand5</th>\n",
       "      <th>Zfp518a</th>\n",
       "      <th>Zfyve27</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.300056</td>\n",
       "      <td>0.209356</td>\n",
       "      <td>0.556247</td>\n",
       "      <td>0.023802</td>\n",
       "      <td>0.003552</td>\n",
       "      <td>0.054591</td>\n",
       "      <td>0.045868</td>\n",
       "      <td>0.108438</td>\n",
       "      <td>0.007118</td>\n",
       "      <td>0.141504</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034883</td>\n",
       "      <td>0.077727</td>\n",
       "      <td>0.045228</td>\n",
       "      <td>0.015388</td>\n",
       "      <td>0.472990</td>\n",
       "      <td>0.314311</td>\n",
       "      <td>0.549204</td>\n",
       "      <td>0.617297</td>\n",
       "      <td>0.208914</td>\n",
       "      <td>0.055166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.113394</td>\n",
       "      <td>0.131442</td>\n",
       "      <td>0.388996</td>\n",
       "      <td>0.014259</td>\n",
       "      <td>0.009897</td>\n",
       "      <td>0.028465</td>\n",
       "      <td>0.005029</td>\n",
       "      <td>0.124987</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004213</td>\n",
       "      <td>0.010186</td>\n",
       "      <td>0.027929</td>\n",
       "      <td>0.003617</td>\n",
       "      <td>0.394133</td>\n",
       "      <td>0.203649</td>\n",
       "      <td>0.348373</td>\n",
       "      <td>0.635299</td>\n",
       "      <td>0.075427</td>\n",
       "      <td>0.020556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.220663</td>\n",
       "      <td>0.160313</td>\n",
       "      <td>0.513679</td>\n",
       "      <td>0.055661</td>\n",
       "      <td>0.004354</td>\n",
       "      <td>0.056418</td>\n",
       "      <td>0.038450</td>\n",
       "      <td>0.134707</td>\n",
       "      <td>0.011689</td>\n",
       "      <td>0.172516</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027235</td>\n",
       "      <td>0.131581</td>\n",
       "      <td>0.039064</td>\n",
       "      <td>0.030639</td>\n",
       "      <td>0.285376</td>\n",
       "      <td>0.255032</td>\n",
       "      <td>0.518535</td>\n",
       "      <td>0.620260</td>\n",
       "      <td>0.269424</td>\n",
       "      <td>0.056064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.282691</td>\n",
       "      <td>0.219056</td>\n",
       "      <td>0.563388</td>\n",
       "      <td>0.049449</td>\n",
       "      <td>0.003716</td>\n",
       "      <td>0.061247</td>\n",
       "      <td>0.043469</td>\n",
       "      <td>0.178020</td>\n",
       "      <td>0.010452</td>\n",
       "      <td>0.171744</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004464</td>\n",
       "      <td>0.076030</td>\n",
       "      <td>0.030168</td>\n",
       "      <td>0.019030</td>\n",
       "      <td>0.443839</td>\n",
       "      <td>0.306142</td>\n",
       "      <td>0.555976</td>\n",
       "      <td>0.754647</td>\n",
       "      <td>0.180463</td>\n",
       "      <td>0.047292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.055400</td>\n",
       "      <td>0.188723</td>\n",
       "      <td>0.693724</td>\n",
       "      <td>0.001685</td>\n",
       "      <td>0.001083</td>\n",
       "      <td>0.007624</td>\n",
       "      <td>0.033985</td>\n",
       "      <td>0.147624</td>\n",
       "      <td>0.001150</td>\n",
       "      <td>0.233223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018612</td>\n",
       "      <td>0.067952</td>\n",
       "      <td>0.080168</td>\n",
       "      <td>0.001150</td>\n",
       "      <td>0.847592</td>\n",
       "      <td>0.315743</td>\n",
       "      <td>0.560970</td>\n",
       "      <td>0.596525</td>\n",
       "      <td>0.107739</td>\n",
       "      <td>0.054099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202_FACS 3m</th>\n",
       "      <td>1.647347</td>\n",
       "      <td>57.811418</td>\n",
       "      <td>6.419415</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029661</td>\n",
       "      <td>22.770750</td>\n",
       "      <td>0.393736</td>\n",
       "      <td>1.650752</td>\n",
       "      <td>2.679234</td>\n",
       "      <td>3.148005</td>\n",
       "      <td>...</td>\n",
       "      <td>1.952788</td>\n",
       "      <td>4.977180</td>\n",
       "      <td>1.770301</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.889174</td>\n",
       "      <td>2.012200</td>\n",
       "      <td>3.277349</td>\n",
       "      <td>44.743875</td>\n",
       "      <td>1.312161</td>\n",
       "      <td>2.156048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203_FACS 3m</th>\n",
       "      <td>0.451712</td>\n",
       "      <td>104.592567</td>\n",
       "      <td>7.415240</td>\n",
       "      <td>0.189706</td>\n",
       "      <td>37.742908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.123764</td>\n",
       "      <td>1.005545</td>\n",
       "      <td>0.066281</td>\n",
       "      <td>6.014079</td>\n",
       "      <td>...</td>\n",
       "      <td>0.443131</td>\n",
       "      <td>1.796151</td>\n",
       "      <td>1.298049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.356874</td>\n",
       "      <td>1.672901</td>\n",
       "      <td>39.011183</td>\n",
       "      <td>15.923136</td>\n",
       "      <td>1.080232</td>\n",
       "      <td>1.315576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204_FACS 3m</th>\n",
       "      <td>2.412300</td>\n",
       "      <td>4.724673</td>\n",
       "      <td>3.808945</td>\n",
       "      <td>11.394566</td>\n",
       "      <td>0.406600</td>\n",
       "      <td>0.871040</td>\n",
       "      <td>0.317492</td>\n",
       "      <td>0.246965</td>\n",
       "      <td>1.233321</td>\n",
       "      <td>1.001302</td>\n",
       "      <td>...</td>\n",
       "      <td>0.887595</td>\n",
       "      <td>1.243274</td>\n",
       "      <td>0.297278</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.372860</td>\n",
       "      <td>0.835273</td>\n",
       "      <td>2.193171</td>\n",
       "      <td>15.958341</td>\n",
       "      <td>0.460870</td>\n",
       "      <td>0.993284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205_FACS 3m</th>\n",
       "      <td>0.645364</td>\n",
       "      <td>30.560482</td>\n",
       "      <td>3.251947</td>\n",
       "      <td>0.032291</td>\n",
       "      <td>0.215574</td>\n",
       "      <td>0.006956</td>\n",
       "      <td>0.309581</td>\n",
       "      <td>0.443338</td>\n",
       "      <td>0.340030</td>\n",
       "      <td>3.436603</td>\n",
       "      <td>...</td>\n",
       "      <td>1.529804</td>\n",
       "      <td>1.978520</td>\n",
       "      <td>0.404200</td>\n",
       "      <td>0.018079</td>\n",
       "      <td>1.573855</td>\n",
       "      <td>0.241060</td>\n",
       "      <td>1.953381</td>\n",
       "      <td>7.341559</td>\n",
       "      <td>0.559050</td>\n",
       "      <td>1.117233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206_FACS 3m</th>\n",
       "      <td>0.686960</td>\n",
       "      <td>6.817909</td>\n",
       "      <td>2.166552</td>\n",
       "      <td>0.450453</td>\n",
       "      <td>0.677547</td>\n",
       "      <td>0.455055</td>\n",
       "      <td>0.070184</td>\n",
       "      <td>0.348857</td>\n",
       "      <td>0.533592</td>\n",
       "      <td>1.306482</td>\n",
       "      <td>...</td>\n",
       "      <td>1.496364</td>\n",
       "      <td>1.279544</td>\n",
       "      <td>0.368024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.569939</td>\n",
       "      <td>0.733703</td>\n",
       "      <td>1.054236</td>\n",
       "      <td>6.431439</td>\n",
       "      <td>0.499511</td>\n",
       "      <td>0.586637</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1206 rows × 11675 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Abi2       Acadl    Actr1b     Adam23     Adhfe1  \\\n",
       "1_E6.5_8.5_Chan  0.300056    0.209356  0.556247   0.023802   0.003552   \n",
       "2_E6.5_8.5_Chan  0.113394    0.131442  0.388996   0.014259   0.009897   \n",
       "3_E6.5_8.5_Chan  0.220663    0.160313  0.513679   0.055661   0.004354   \n",
       "4_E6.5_8.5_Chan  0.282691    0.219056  0.563388   0.049449   0.003716   \n",
       "5_E6.5_8.5_Chan  0.055400    0.188723  0.693724   0.001685   0.001083   \n",
       "...                   ...         ...       ...        ...        ...   \n",
       "1202_FACS 3m     1.647347   57.811418  6.419415   0.000000   0.029661   \n",
       "1203_FACS 3m     0.451712  104.592567  7.415240   0.189706  37.742908   \n",
       "1204_FACS 3m     2.412300    4.724673  3.808945  11.394566   0.406600   \n",
       "1205_FACS 3m     0.645364   30.560482  3.251947   0.032291   0.215574   \n",
       "1206_FACS 3m     0.686960    6.817909  2.166552   0.450453   0.677547   \n",
       "\n",
       "                      Aff3       Als2   Ankrd39   Ankrd44   Arfgef1  ...  \\\n",
       "1_E6.5_8.5_Chan   0.054591   0.045868  0.108438  0.007118  0.141504  ...   \n",
       "2_E6.5_8.5_Chan   0.028465   0.005029  0.124987  0.000000  0.065782  ...   \n",
       "3_E6.5_8.5_Chan   0.056418   0.038450  0.134707  0.011689  0.172516  ...   \n",
       "4_E6.5_8.5_Chan   0.061247   0.043469  0.178020  0.010452  0.171744  ...   \n",
       "5_E6.5_8.5_Chan   0.007624   0.033985  0.147624  0.001150  0.233223  ...   \n",
       "...                    ...        ...       ...       ...       ...  ...   \n",
       "1202_FACS 3m     22.770750   0.393736  1.650752  2.679234  3.148005  ...   \n",
       "1203_FACS 3m      0.000000  11.123764  1.005545  0.066281  6.014079  ...   \n",
       "1204_FACS 3m      0.871040   0.317492  0.246965  1.233321  1.001302  ...   \n",
       "1205_FACS 3m      0.006956   0.309581  0.443338  0.340030  3.436603  ...   \n",
       "1206_FACS 3m      0.455055   0.070184  0.348857  0.533592  1.306482  ...   \n",
       "\n",
       "                    Vldlr    Vps13a     Vti1a      Vwa2   Xpnpep1   Zdhhc16  \\\n",
       "1_E6.5_8.5_Chan  0.034883  0.077727  0.045228  0.015388  0.472990  0.314311   \n",
       "2_E6.5_8.5_Chan  0.004213  0.010186  0.027929  0.003617  0.394133  0.203649   \n",
       "3_E6.5_8.5_Chan  0.027235  0.131581  0.039064  0.030639  0.285376  0.255032   \n",
       "4_E6.5_8.5_Chan  0.004464  0.076030  0.030168  0.019030  0.443839  0.306142   \n",
       "5_E6.5_8.5_Chan  0.018612  0.067952  0.080168  0.001150  0.847592  0.315743   \n",
       "...                   ...       ...       ...       ...       ...       ...   \n",
       "1202_FACS 3m     1.952788  4.977180  1.770301  0.000000  2.889174  2.012200   \n",
       "1203_FACS 3m     0.443131  1.796151  1.298049  0.000000  6.356874  1.672901   \n",
       "1204_FACS 3m     0.887595  1.243274  0.297278  0.000000  2.372860  0.835273   \n",
       "1205_FACS 3m     1.529804  1.978520  0.404200  0.018079  1.573855  0.241060   \n",
       "1206_FACS 3m     1.496364  1.279544  0.368024  0.000000  1.569939  0.733703   \n",
       "\n",
       "                    Zdhhc6     Zfand5   Zfp518a   Zfyve27  \n",
       "1_E6.5_8.5_Chan   0.549204   0.617297  0.208914  0.055166  \n",
       "2_E6.5_8.5_Chan   0.348373   0.635299  0.075427  0.020556  \n",
       "3_E6.5_8.5_Chan   0.518535   0.620260  0.269424  0.056064  \n",
       "4_E6.5_8.5_Chan   0.555976   0.754647  0.180463  0.047292  \n",
       "5_E6.5_8.5_Chan   0.560970   0.596525  0.107739  0.054099  \n",
       "...                    ...        ...       ...       ...  \n",
       "1202_FACS 3m      3.277349  44.743875  1.312161  2.156048  \n",
       "1203_FACS 3m     39.011183  15.923136  1.080232  1.315576  \n",
       "1204_FACS 3m      2.193171  15.958341  0.460870  0.993284  \n",
       "1205_FACS 3m      1.953381   7.341559  0.559050  1.117233  \n",
       "1206_FACS 3m      1.054236   6.431439  0.499511  0.586637  \n",
       "\n",
       "[1206 rows x 11675 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data Normalization <a name=\"norm\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The imported count dataset contains normalized counts after processing from each dataset. However, to eliminate differences in factors like sequencing depth (from different sequencing protocols) and cell size (between different cells), we can normalize the pseudo-bulk sample counts so that they have the same total counts per sample, 1e4. \n",
    "\n",
    "Additionally, in scRNA-seq analysis, many downstream computations on count data typically assume normality of their distributions. However, raw count data from scRNA-seq is highly skewed, and we conduct a log-plus-one normalization to bring the data \"closer\" to normality. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1_E6.5_8.5_Chan    10000.0\n",
       "2_E6.5_8.5_Chan    10000.0\n",
       "3_E6.5_8.5_Chan    10000.0\n",
       "4_E6.5_8.5_Chan    10000.0\n",
       "5_E6.5_8.5_Chan    10000.0\n",
       "                    ...   \n",
       "1202_FACS 3m       10000.0\n",
       "1203_FACS 3m       10000.0\n",
       "1204_FACS 3m       10000.0\n",
       "1205_FACS 3m       10000.0\n",
       "1206_FACS 3m       10000.0\n",
       "Length: 1206, dtype: float64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalize the data such that the total counts per cluster is 1e4, then take the log\n",
    "integrated_counts = np.log1p(counts.div(counts.sum(axis=1),axis=0)*1e4)\n",
    "\n",
    "# A quick check: the total counts for each pseudo-bulk sample indeed add to 1e4\n",
    "np.expm1(integrated_counts).sum(axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plots<a name=\"plots\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BMP Receptors<a name=\"bmpr\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have MinMax scaled gene expression, and can probe expression of genes in specific pathways of interest and evaluate how these vary across different timepoints, cell types, etc. \n",
    "\n",
    "I'll demonstrate some possible queries we can make from this atlas dataset using the Notch signaling pathway. \n",
    "\n",
    "Note, we can substitute this pathway with any other pathways or subsets of genes. The DataFrame \"pathways\" contains a curated list of genes in select pathways (shown below). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Bmp_l', 'Notch', 'Bmp_Tgfb', 'Bmp', 'Wnt', 'Eph_r', 'Eph_l',\n",
       "       'Fgfr', 'Srsf', 'Lpa', 'Wnt_l', 'Bmp_down', 'Bmp_fun'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The pathways that are in our curated list\n",
    "# l means ligands, r means receptors\n",
    "pathways[\"pathway\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will use the list of Notch components stored in our curated list above. Otherwise, if you want to study a set of signaling genes not in this csv, simply change the pathway name and list of genes below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the Notch genes from our reference list\n",
    "pathway = \"Bmp_Tgfb\"\n",
    "genes = list(pathways[pathways[\"pathway\"] == pathway][\"gene\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As of now, the integrated atlas only consists of the intersection of genes that are expressed at a basal level in all datasets (for a total of 11,675 genes out of ~22k total). Some genes specified in our list above may not be contained within this intersection, so we filter them out below, and subset the remaining pathway genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bmpr1a</th>\n",
       "      <th>Bmpr1b</th>\n",
       "      <th>Acvr1</th>\n",
       "      <th>Acvrl1</th>\n",
       "      <th>Acvr1b</th>\n",
       "      <th>Tgfbr1</th>\n",
       "      <th>Acvr1c</th>\n",
       "      <th>Acvr2a</th>\n",
       "      <th>Acvr2b</th>\n",
       "      <th>Bmpr2</th>\n",
       "      <th>Tgfbr2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.259442</td>\n",
       "      <td>0.009422</td>\n",
       "      <td>0.039890</td>\n",
       "      <td>0.003299</td>\n",
       "      <td>0.054732</td>\n",
       "      <td>0.105739</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029469</td>\n",
       "      <td>0.460605</td>\n",
       "      <td>0.102073</td>\n",
       "      <td>0.001512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.082582</td>\n",
       "      <td>0.007682</td>\n",
       "      <td>0.009354</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005502</td>\n",
       "      <td>0.029284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.127896</td>\n",
       "      <td>0.006597</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.294106</td>\n",
       "      <td>0.013823</td>\n",
       "      <td>0.031756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.063601</td>\n",
       "      <td>0.099238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037264</td>\n",
       "      <td>0.503437</td>\n",
       "      <td>0.060422</td>\n",
       "      <td>0.005415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.269606</td>\n",
       "      <td>0.010418</td>\n",
       "      <td>0.033919</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058188</td>\n",
       "      <td>0.131021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.034061</td>\n",
       "      <td>0.510978</td>\n",
       "      <td>0.082296</td>\n",
       "      <td>0.005668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.146207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013514</td>\n",
       "      <td>0.018657</td>\n",
       "      <td>0.064921</td>\n",
       "      <td>0.061244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.067876</td>\n",
       "      <td>0.325515</td>\n",
       "      <td>0.042243</td>\n",
       "      <td>0.020734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Bmpr1a    Bmpr1b     Acvr1    Acvrl1    Acvr1b    Tgfbr1  \\\n",
       "1_E6.5_8.5_Chan  0.259442  0.009422  0.039890  0.003299  0.054732  0.105739   \n",
       "2_E6.5_8.5_Chan  0.082582  0.007682  0.009354  0.000000  0.005502  0.029284   \n",
       "3_E6.5_8.5_Chan  0.294106  0.013823  0.031756  0.000000  0.063601  0.099238   \n",
       "4_E6.5_8.5_Chan  0.269606  0.010418  0.033919  0.000000  0.058188  0.131021   \n",
       "5_E6.5_8.5_Chan  0.146207  0.000000  0.013514  0.018657  0.064921  0.061244   \n",
       "\n",
       "                 Acvr1c    Acvr2a    Acvr2b     Bmpr2    Tgfbr2  \n",
       "1_E6.5_8.5_Chan     0.0  0.029469  0.460605  0.102073  0.001512  \n",
       "2_E6.5_8.5_Chan     0.0  0.011628  0.127896  0.006597  0.000000  \n",
       "3_E6.5_8.5_Chan     0.0  0.037264  0.503437  0.060422  0.005415  \n",
       "4_E6.5_8.5_Chan     0.0  0.034061  0.510978  0.082296  0.005668  \n",
       "5_E6.5_8.5_Chan     0.0  0.067876  0.325515  0.042243  0.020734  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset a DataFrame with only the genes selected above\n",
    "pathway_genes = [i for i in genes if i in integrated_counts.columns]\n",
    "\n",
    "df_pathway = integrated_counts[pathway_genes]\n",
    "\n",
    "df_pathway.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Scaling of Counts <a name=\"scaling\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compare expression of different genes in a pathway, we need to account for the the fact that counts for different genes may have very different dynamic ranges, and any direct clustering on these normalized data would be biased toward highly-expressed genes. To give equal weight to all features, we can perform MinMax scaling to scale counts per gene to lie in the range of 0 to 1. However, since RNA count distributions are right-tailed, scaling the maximum count value per gene to correspond to a value of 1 would sharply skew the new, scaled. values. We can instead more robustly saturate count values by scaling the count value that corresponds to a scaled expression value of 1.0 to be some percentile of maximum gene expression (i.e. the 99th percentile). \n",
    "\n",
    "The function below saturates the expression values in a DataFrame at inputted gene-specific percentiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Returns a clipped matrix given a list of max values (values to clip at) for each gene\n",
    "def quantile_saturation(x = [], gene_quantiles = []):\n",
    "    \n",
    "    col_order = list(x.columns)\n",
    "    \n",
    "    x_clipped = pd.DataFrame()\n",
    "    # Saturate the values based on 99% quantile (more robust than using the abs max)\n",
    "    for i in x.index:\n",
    "        x_clipped =  x_clipped.append(np.clip(x.loc[i], a_min=0, a_max = gene_quantiles))\n",
    "        \n",
    "    # Pandas automatically makes the new df columns alphabetically ordered\n",
    "    # re-order columns how they were originally entered by the user\n",
    "    \n",
    "    x_clipped = x_clipped[col_order]\n",
    "    \n",
    "    return x_clipped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bmpr1a</th>\n",
       "      <th>Bmpr1b</th>\n",
       "      <th>Acvr1</th>\n",
       "      <th>Acvrl1</th>\n",
       "      <th>Acvr1b</th>\n",
       "      <th>Tgfbr1</th>\n",
       "      <th>Acvr1c</th>\n",
       "      <th>Acvr2a</th>\n",
       "      <th>Acvr2b</th>\n",
       "      <th>Bmpr2</th>\n",
       "      <th>Tgfbr2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.174746</td>\n",
       "      <td>0.010185</td>\n",
       "      <td>0.047448</td>\n",
       "      <td>0.207868</td>\n",
       "      <td>0.019365</td>\n",
       "      <td>0.204041</td>\n",
       "      <td>0.003809</td>\n",
       "      <td>0.056347</td>\n",
       "      <td>0.171019</td>\n",
       "      <td>0.077264</td>\n",
       "      <td>0.253857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.167259</td>\n",
       "      <td>0.028057</td>\n",
       "      <td>0.056113</td>\n",
       "      <td>0.451051</td>\n",
       "      <td>0.027337</td>\n",
       "      <td>0.216681</td>\n",
       "      <td>0.017101</td>\n",
       "      <td>0.052769</td>\n",
       "      <td>0.227145</td>\n",
       "      <td>0.110175</td>\n",
       "      <td>0.319152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.024696</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007072</td>\n",
       "      <td>0.003779</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071994</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016728</td>\n",
       "      <td>0.008508</td>\n",
       "      <td>0.009665</td>\n",
       "      <td>0.010673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.128166</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026510</td>\n",
       "      <td>0.023229</td>\n",
       "      <td>0.007058</td>\n",
       "      <td>0.136030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041313</td>\n",
       "      <td>0.051766</td>\n",
       "      <td>0.033868</td>\n",
       "      <td>0.129044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.280556</td>\n",
       "      <td>0.006436</td>\n",
       "      <td>0.065708</td>\n",
       "      <td>0.172660</td>\n",
       "      <td>0.028376</td>\n",
       "      <td>0.275623</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.078793</td>\n",
       "      <td>0.256899</td>\n",
       "      <td>0.088366</td>\n",
       "      <td>0.376579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.647956</td>\n",
       "      <td>0.180026</td>\n",
       "      <td>0.257370</td>\n",
       "      <td>2.936467</td>\n",
       "      <td>0.138782</td>\n",
       "      <td>1.407423</td>\n",
       "      <td>0.126763</td>\n",
       "      <td>0.250110</td>\n",
       "      <td>0.799456</td>\n",
       "      <td>0.521438</td>\n",
       "      <td>1.417466</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Bmpr1a       Bmpr1b        Acvr1       Acvrl1       Acvr1b  \\\n",
       "count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   \n",
       "mean      0.174746     0.010185     0.047448     0.207868     0.019365   \n",
       "std       0.167259     0.028057     0.056113     0.451051     0.027337   \n",
       "min       0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "25%       0.024696     0.000000     0.007072     0.003779     0.000000   \n",
       "50%       0.128166     0.000000     0.026510     0.023229     0.007058   \n",
       "75%       0.280556     0.006436     0.065708     0.172660     0.028376   \n",
       "max       0.647956     0.180026     0.257370     2.936467     0.138782   \n",
       "\n",
       "            Tgfbr1       Acvr1c       Acvr2a       Acvr2b        Bmpr2  \\\n",
       "count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   \n",
       "mean      0.204041     0.003809     0.056347     0.171019     0.077264   \n",
       "std       0.216681     0.017101     0.052769     0.227145     0.110175   \n",
       "min       0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "25%       0.071994     0.000000     0.016728     0.008508     0.009665   \n",
       "50%       0.136030     0.000000     0.041313     0.051766     0.033868   \n",
       "75%       0.275623     0.000077     0.078793     0.256899     0.088366   \n",
       "max       1.407423     0.126763     0.250110     0.799456     0.521438   \n",
       "\n",
       "            Tgfbr2  \n",
       "count  1206.000000  \n",
       "mean      0.253857  \n",
       "std       0.319152  \n",
       "min       0.000000  \n",
       "25%       0.010673  \n",
       "50%       0.129044  \n",
       "75%       0.376579  \n",
       "max       1.417466  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute the 99th quantile per gene\n",
    "gene_quantiles = np.quantile(df_pathway, q = 0.99, axis = 0)\n",
    "\n",
    "# Clip the values such that any counts above the 99th quantile take its value\n",
    "x_clipped = quantile_saturation(df_pathway, gene_quantiles)\n",
    "x_clipped.describe() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now scale expression values in this DataFrame of \"clipped\" counts to lie between 0 and 1, using a simple MinMaxScaler from scikitlearn that linearly scales counts to lie between the a specified minimum and maximum value per gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bmpr1a</th>\n",
       "      <th>Bmpr1b</th>\n",
       "      <th>Acvr1</th>\n",
       "      <th>Acvrl1</th>\n",
       "      <th>Acvr1b</th>\n",
       "      <th>Tgfbr1</th>\n",
       "      <th>Acvr1c</th>\n",
       "      <th>Acvr2a</th>\n",
       "      <th>Acvr2b</th>\n",
       "      <th>Bmpr2</th>\n",
       "      <th>Tgfbr2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.400400</td>\n",
       "      <td>0.052337</td>\n",
       "      <td>0.154991</td>\n",
       "      <td>0.001123</td>\n",
       "      <td>0.394376</td>\n",
       "      <td>0.075130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.117825</td>\n",
       "      <td>0.576148</td>\n",
       "      <td>0.195752</td>\n",
       "      <td>0.001067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.127450</td>\n",
       "      <td>0.042669</td>\n",
       "      <td>0.036343</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.039646</td>\n",
       "      <td>0.020807</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.046493</td>\n",
       "      <td>0.159978</td>\n",
       "      <td>0.012652</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.453898</td>\n",
       "      <td>0.076784</td>\n",
       "      <td>0.123388</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.458279</td>\n",
       "      <td>0.070510</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.148989</td>\n",
       "      <td>0.629725</td>\n",
       "      <td>0.115875</td>\n",
       "      <td>0.003820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.416087</td>\n",
       "      <td>0.057867</td>\n",
       "      <td>0.131791</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.419273</td>\n",
       "      <td>0.093093</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.136184</td>\n",
       "      <td>0.639157</td>\n",
       "      <td>0.157825</td>\n",
       "      <td>0.003999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.225643</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.052509</td>\n",
       "      <td>0.006353</td>\n",
       "      <td>0.467789</td>\n",
       "      <td>0.043515</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.271386</td>\n",
       "      <td>0.407170</td>\n",
       "      <td>0.081013</td>\n",
       "      <td>0.014628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Bmpr1a    Bmpr1b     Acvr1    Acvrl1    Acvr1b    Tgfbr1  \\\n",
       "1_E6.5_8.5_Chan  0.400400  0.052337  0.154991  0.001123  0.394376  0.075130   \n",
       "2_E6.5_8.5_Chan  0.127450  0.042669  0.036343  0.000000  0.039646  0.020807   \n",
       "3_E6.5_8.5_Chan  0.453898  0.076784  0.123388  0.000000  0.458279  0.070510   \n",
       "4_E6.5_8.5_Chan  0.416087  0.057867  0.131791  0.000000  0.419273  0.093093   \n",
       "5_E6.5_8.5_Chan  0.225643  0.000000  0.052509  0.006353  0.467789  0.043515   \n",
       "\n",
       "                 Acvr1c    Acvr2a    Acvr2b     Bmpr2    Tgfbr2  \n",
       "1_E6.5_8.5_Chan     0.0  0.117825  0.576148  0.195752  0.001067  \n",
       "2_E6.5_8.5_Chan     0.0  0.046493  0.159978  0.012652  0.000000  \n",
       "3_E6.5_8.5_Chan     0.0  0.148989  0.629725  0.115875  0.003820  \n",
       "4_E6.5_8.5_Chan     0.0  0.136184  0.639157  0.157825  0.003999  \n",
       "5_E6.5_8.5_Chan     0.0  0.271386  0.407170  0.081013  0.014628  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize the scaler\n",
    "scaler = MinMaxScaler()\n",
    "# Fit the scaler to the dataset to compute per-feature min. and max. values\n",
    "scaler.fit(x_clipped)\n",
    "\n",
    "# Apply the scaler to the dataset\n",
    "df_scale = pd.DataFrame(scaler.transform(x_clipped), columns = x_clipped.columns, index = x_clipped.index)\n",
    "\n",
    "df_scale.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BMPR Correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at gene correlation for clusters that have the BMP receptors active. We can try to measure this by first computing correlation only at clusters that have expression of 2 receptors greater than or equal to 0.2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(806, 11)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filt = pd.DataFrame(columns = df_scale.columns)\n",
    "\n",
    "for i in df_scale.index:\n",
    "    if (df_scale.loc[i][pathway_genes] >= 0.2).sum() > 1:\n",
    "        df_filt = df_filt.append(df_scale.loc[i])\n",
    "        \n",
    "df_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_corr = as.data.frame(cor(df_filt))\n",
    "colnames(M_corr) = colnames(df_filt)\n",
    "rownames(M_corr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/bmpr_corrplot_all.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_corr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BMPR Partial Correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_pcorr = as.data.frame(pcor(df_filt)$estimate)\n",
    "colnames(M_pcorr) = colnames(df_filt)\n",
    "rownames(M_pcorr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/bmpr_parcorrplot_all.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_pcorr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%R -o M_corr,M_pcorr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_corr = M_corr.where(np.triu(np.ones(M_corr.shape), k = 1).astype(np.bool))\n",
    "M_pcorr = M_pcorr.where(np.triu(np.ones(M_pcorr.shape), k = 1).astype(np.bool))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_max = np.zeros((2,2))\n",
    "\n",
    "min_max[0] = min(np.nanmin(M_corr.min(skipna = True)), np.nanmin(M_pcorr.min(skipna = True)))\n",
    "min_max[1] = max(np.nanmax(M_corr.max(skipna = True)), np.nanmax(M_pcorr.max(skipna = True)))\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for i in list(itertools.combinations(df_filt.columns, 2)):\n",
    "    x = pd.DataFrame()\n",
    "    df = df.append([[M_corr.loc[i[0]][i[1]], M_pcorr.loc[i[0]][i[1]], i]])\n",
    "    \n",
    "df.columns = [\"corr\", \"pcorr\", \"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"f30c4033-9855-46e6-b318-c5fedc598e7d\" data-root-id=\"1009\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"3601adfa-4d8e-4b46-b412-7bc4472787c6\":{\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"1020\"}],\"center\":[{\"id\":\"1023\"},{\"id\":\"1027\"}],\"left\":[{\"id\":\"1024\"}],\"plot_height\":400,\"plot_width\":400,\"renderers\":[{\"id\":\"1039\"},{\"id\":\"1044\"}],\"title\":{\"id\":\"1010\"},\"toolbar\":{\"id\":\"1029\"},\"toolbar_location\":\"below\",\"x_range\":{\"id\":\"1012\"},\"x_scale\":{\"id\":\"1016\"},\"y_range\":{\"id\":\"1014\"},\"y_scale\":{\"id\":\"1018\"}},\"id\":\"1009\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"axis_label\":\"Correlation\",\"formatter\":{\"id\":\"1050\"},\"ticker\":{\"id\":\"1021\"}},\"id\":\"1020\",\"type\":\"LinearAxis\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"1003\"},{\"id\":\"1004\"},{\"id\":\"1005\"},{\"id\":\"1006\"},{\"id\":\"1007\"},{\"id\":\"1008\"}]},\"id\":\"1029\",\"type\":\"Toolbar\"},{\"attributes\":{\"data_source\":{\"id\":\"1002\"},\"glyph\":{\"id\":\"1037\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1038\"},\"selection_glyph\":null,\"view\":{\"id\":\"1040\"}},\"id\":\"1039\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"1018\",\"type\":\"LinearScale\"},{\"attributes\":{\"data_source\":{\"id\":\"1041\"},\"glyph\":{\"id\":\"1042\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1043\"},\"selection_glyph\":null,\"view\":{\"id\":\"1045\"}},\"id\":\"1044\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"1025\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"1050\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"axis\":{\"id\":\"1024\"},\"dimension\":1,\"ticker\":null},\"id\":\"1027\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"1021\",\"type\":\"BasicTicker\"},{\"attributes\":{\"axis_label\":\"Partial Correlation\",\"formatter\":{\"id\":\"1048\"},\"ticker\":{\"id\":\"1025\"}},\"id\":\"1024\",\"type\":\"LinearAxis\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1038\",\"type\":\"Scatter\"},{\"attributes\":{\"fill_color\":{\"value\":\"#1f77b4\"},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1037\",\"type\":\"Scatter\"},{\"attributes\":{},\"id\":\"1016\",\"type\":\"LinearScale\"},{\"attributes\":{},\"id\":\"1014\",\"type\":\"DataRange1d\"},{\"attributes\":{},\"id\":\"1052\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1048\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"data\":{\"corr\":{\"__ndarray__\":\"MDw5ZjnR2z+vxLVXb2XYP7d+PWU4z9O/E9uh/b+Gzz9qzZza9bmlPwaqAbisU4A/D4tu0IZI1T8vMHMkBj3PP+FQk4ulGNM/UO1gm/Dpxr/5lOo3cibHP+X0/zsDKsS/+I+RwgCBuD+d+6b/E+jKP1jC/6oPxsc//TnLfiys1D+7rAqH9HOiPyTGYqBdOdY/nriHn9BruL+xavWVZb3HP9wyY2DGncK/op0/yRa4xz98RbwESiK4PyA3F9ek5tY/7CtqQejb0r+iTupijizWP0ljXBo5ddk/SOm0ZFib0b8G78S2ZSNzvwt+U8wPabK/3Ot34G1IvT8PXoQrZLHTvwXpgQD6GtE/5Z18xbL73T8hAZXN5JbGv7LgKbuF5pK/bezdDo7qlL/Xq9F1oLncP+b5EbqcTs0/uAOfZW9Z278SVWPZgKSRv799dCXFDqo/rC/0Sq8TzL+3Q0kqQHi4P8JuMsFR7Nk/ZKvb30ofzj99ea+rfWvAvwlSr7/3S7a/SvXEBng4pL8O9+WwmBHDv3/jG0tmVNc/KFINyFEEyD8NkJEr9I1rPyAhnUK3bOG/+jmz7xCVnT8=\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[55]},\"id\":[[\"Bmpr1a\",\"Bmpr1b\"],[\"Bmpr1a\",\"Acvr1\"],[\"Bmpr1a\",\"Acvrl1\"],[\"Bmpr1a\",\"Acvr1b\"],[\"Bmpr1a\",\"Tgfbr1\"],[\"Bmpr1a\",\"Acvr1c\"],[\"Bmpr1a\",\"Acvr2a\"],[\"Bmpr1a\",\"Acvr2b\"],[\"Bmpr1a\",\"Bmpr2\"],[\"Bmpr1a\",\"Tgfbr2\"],[\"Bmpr1b\",\"Acvr1\"],[\"Bmpr1b\",\"Acvrl1\"],[\"Bmpr1b\",\"Acvr1b\"],[\"Bmpr1b\",\"Tgfbr1\"],[\"Bmpr1b\",\"Acvr1c\"],[\"Bmpr1b\",\"Acvr2a\"],[\"Bmpr1b\",\"Acvr2b\"],[\"Bmpr1b\",\"Bmpr2\"],[\"Bmpr1b\",\"Tgfbr2\"],[\"Acvr1\",\"Acvrl1\"],[\"Acvr1\",\"Acvr1b\"],[\"Acvr1\",\"Tgfbr1\"],[\"Acvr1\",\"Acvr1c\"],[\"Acvr1\",\"Acvr2a\"],[\"Acvr1\",\"Acvr2b\"],[\"Acvr1\",\"Bmpr2\"],[\"Acvr1\",\"Tgfbr2\"],[\"Acvrl1\",\"Acvr1b\"],[\"Acvrl1\",\"Tgfbr1\"],[\"Acvrl1\",\"Acvr1c\"],[\"Acvrl1\",\"Acvr2a\"],[\"Acvrl1\",\"Acvr2b\"],[\"Acvrl1\",\"Bmpr2\"],[\"Acvrl1\",\"Tgfbr2\"],[\"Acvr1b\",\"Tgfbr1\"],[\"Acvr1b\",\"Acvr1c\"],[\"Acvr1b\",\"Acvr2a\"],[\"Acvr1b\",\"Acvr2b\"],[\"Acvr1b\",\"Bmpr2\"],[\"Acvr1b\",\"Tgfbr2\"],[\"Tgfbr1\",\"Acvr1c\"],[\"Tgfbr1\",\"Acvr2a\"],[\"Tgfbr1\",\"Acvr2b\"],[\"Tgfbr1\",\"Bmpr2\"],[\"Tgfbr1\",\"Tgfbr2\"],[\"Acvr1c\",\"Acvr2a\"],[\"Acvr1c\",\"Acvr2b\"],[\"Acvr1c\",\"Bmpr2\"],[\"Acvr1c\",\"Tgfbr2\"],[\"Acvr2a\",\"Acvr2b\"],[\"Acvr2a\",\"Bmpr2\"],[\"Acvr2a\",\"Tgfbr2\"],[\"Acvr2b\",\"Bmpr2\"],[\"Acvr2b\",\"Tgfbr2\"],[\"Bmpr2\",\"Tgfbr2\"]],\"index\":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],\"pcorr\":{\"__ndarray__\":\"c0lR4DxR0T8Rt7p8r2TcP8iVUjIY69G/AxS42lfSsz+0dmzxuip7v51htVrxvsK/Pn7Sf8AozT8avH72yBnNP6iwcMJfN6k/lW1M9ZpKt78xudG474O1v7k9JTXy2ba/TFCGD3egs7/lPp7P/y/MP9P36n7M1Mg/RaZSj/wcwD86TwAv78Glv3fMwzkUldA/9vCgNRhaub94jhowTiKYP0Oz0TnFAri/3HAg87jnob/fP95SQvbBP04Q+5T2Apg/cFphKO+7xb+r7MNCqUnPP8Fo0nPvOtM/RSYAZafmxb97zfiod5jOv6twJ4ylVI2/DIj11IF5kj+zjNjNqKyYPy/YHWwTTdo/baljlhD31j8EnRqgeaayv779uV3z4rA/jbYoy7RDpb+HUXRBTZjPP9G0rFTiBtQ/g1g1SJ3st79ZfsxXjnSIv7I9R1RixMC/1vXblkLef7+csCY3ZkrBP5KQJYyN69o/0UG96+Vr0D99cR/NAZq/v5h49yEyoMy/sECK9JpHwr/E9f9RsLCsv3n0vUy5Y84/iGFzC3Myxz8K4Mlx9SOhvwZNXj5IBdK/tewAB0aHyL8=\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[55]}},\"selected\":{\"id\":\"1052\"},\"selection_policy\":{\"id\":\"1053\"}},\"id\":\"1002\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"1053\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"axis\":{\"id\":\"1020\"},\"ticker\":null},\"id\":\"1023\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"1008\",\"type\":\"SaveTool\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Genes\",\"@id\"]]},\"id\":\"1003\",\"type\":\"HoverTool\"},{\"attributes\":{\"text\":\"BMP Receptors\"},\"id\":\"1010\",\"type\":\"Title\"},{\"attributes\":{\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1042\",\"type\":\"Line\"},{\"attributes\":{},\"id\":\"1054\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1004\",\"type\":\"WheelZoomTool\"},{\"attributes\":{\"source\":{\"id\":\"1041\"}},\"id\":\"1045\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1012\",\"type\":\"DataRange1d\"},{\"attributes\":{\"line_alpha\":0.1,\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1043\",\"type\":\"Line\"},{\"attributes\":{},\"id\":\"1055\",\"type\":\"UnionRenderers\"},{\"attributes\":{},\"id\":\"1005\",\"type\":\"PanTool\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"1028\",\"type\":\"BoxAnnotation\"},{\"attributes\":{\"source\":{\"id\":\"1002\"}},\"id\":\"1040\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1007\",\"type\":\"ResetTool\"},{\"attributes\":{\"data\":{\"x\":{\"__ndarray__\":\"ICGdQrds4b/lnXzFsvvdPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]},\"y\":{\"__ndarray__\":\"ICGdQrds4b/lnXzFsvvdPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]}},\"selected\":{\"id\":\"1054\"},\"selection_policy\":{\"id\":\"1055\"}},\"id\":\"1041\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"overlay\":{\"id\":\"1028\"}},\"id\":\"1006\",\"type\":\"BoxZoomTool\"}],\"root_ids\":[\"1009\"]},\"title\":\"Bokeh Application\",\"version\":\"2.1.1\"}};\n",
       "  var render_items = [{\"docid\":\"3601adfa-4d8e-4b46-b412-7bc4472787c6\",\"root_ids\":[\"1009\"],\"roots\":{\"1009\":\"f30c4033-9855-46e6-b318-c5fedc598e7d\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "1009"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "source = ColumnDataSource(df)\n",
    "\n",
    "hover = HoverTool(tooltips=[('Genes', '@id')])\n",
    "tools = [hover, WheelZoomTool(), PanTool(), BoxZoomTool(), ResetTool(), SaveTool()]\n",
    "\n",
    "p = figure(width=400, height=400,tools=tools,\n",
    "           toolbar_location=\"below\", \n",
    "           x_axis_label = \"Correlation\",\n",
    "           y_axis_label = \"Partial Correlation\",\n",
    "           title = \"BMP Receptors\")\n",
    "\n",
    "p.scatter(x = \"corr\", y = \"pcorr\", source = source, color = \"#1f77b4\");\n",
    "p.line(min_max[:,0], min_max[:,1], line_color = \"black\");\n",
    "\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at gene correlation for clusters that have the BMP receptors active. We can try to measure this by first computing correlation only at clusters that have expression of at least one Type I and one Type II receptor above a threshold of 0.2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "bmpr_type_one = ['Bmpr1a', 'Bmpr1b', 'Acvr1', 'Acvrl1', 'Acvr1b', 'Tgfbr1', 'Acvr1c']\n",
    "bmpr_type_two = ['Acvr2a', 'Acvr2b', 'Bmpr2', 'Tgfbr2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(756, 11)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filt = pd.DataFrame(columns = df_scale.columns)\n",
    "\n",
    "for i in df_scale.index:\n",
    "    if ((df_scale.loc[i][bmpr_type_one] >= 0.2).any()):\n",
    "        if ((df_scale.loc[i][bmpr_type_two] > 0.2).any()):\n",
    "            df_filt = df_filt.append(df_scale.loc[i])\n",
    "        \n",
    "df_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_corr = as.data.frame(cor(df_filt))\n",
    "colnames(M_pcorr) = colnames(df_filt)\n",
    "rownames(M_pcorr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/bmpr_corrplot_type_i_ii.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_corr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BMPR Partial Correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M = as.data.frame(pcor(df_filt)$estimate)\n",
    "colnames(M_pcorr) = colnames(df_filt)\n",
    "rownames(M_pcorr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/bmpr_parcorrplot_type_i_ii.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_pcorr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "%R -o M_corr,M_pcorr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_corr = M_corr.where(np.triu(np.ones(M_corr.shape), k = 1).astype(np.bool))\n",
    "M_pcorr = M_pcorr.where(np.triu(np.ones(M_pcorr.shape), k = 1).astype(np.bool))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_max = np.zeros((2,2))\n",
    "\n",
    "min_max[0] = min(np.nanmin(M_corr.min(skipna = True)), np.nanmin(M_pcorr.min(skipna = True)))\n",
    "min_max[1] = max(np.nanmax(M_corr.max(skipna = True)), np.nanmax(M_pcorr.max(skipna = True)))\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for i in list(itertools.combinations(df_filt.columns, 2)):\n",
    "    x = pd.DataFrame()\n",
    "    df = df.append([[M_corr.loc[i[0]][i[1]], M_pcorr.loc[i[0]][i[1]], i]])\n",
    "    \n",
    "df.columns = [\"corr\", \"pcorr\", \"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"77c4c74f-5278-49fa-a5b7-d846ac2d35ab\" data-root-id=\"1123\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"159619af-7bd9-4928-9df1-b40a1f61eb47\":{\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"1134\"}],\"center\":[{\"id\":\"1137\"},{\"id\":\"1141\"}],\"left\":[{\"id\":\"1138\"}],\"plot_height\":400,\"plot_width\":400,\"renderers\":[{\"id\":\"1153\"},{\"id\":\"1158\"}],\"title\":{\"id\":\"1124\"},\"toolbar\":{\"id\":\"1143\"},\"toolbar_location\":\"below\",\"x_range\":{\"id\":\"1126\"},\"x_scale\":{\"id\":\"1130\"},\"y_range\":{\"id\":\"1128\"},\"y_scale\":{\"id\":\"1132\"}},\"id\":\"1123\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"text\":\"BMP Receptors\"},\"id\":\"1124\",\"type\":\"Title\"},{\"attributes\":{},\"id\":\"1128\",\"type\":\"DataRange1d\"},{\"attributes\":{\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1156\",\"type\":\"Line\"},{\"attributes\":{\"fill_color\":{\"value\":\"#1f77b4\"},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1151\",\"type\":\"Scatter\"},{\"attributes\":{},\"id\":\"1121\",\"type\":\"ResetTool\"},{\"attributes\":{},\"id\":\"1122\",\"type\":\"SaveTool\"},{\"attributes\":{},\"id\":\"1130\",\"type\":\"LinearScale\"},{\"attributes\":{\"axis\":{\"id\":\"1138\"},\"dimension\":1,\"ticker\":null},\"id\":\"1141\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"1118\",\"type\":\"WheelZoomTool\"},{\"attributes\":{\"axis\":{\"id\":\"1134\"},\"ticker\":null},\"id\":\"1137\",\"type\":\"Grid\"},{\"attributes\":{\"source\":{\"id\":\"1155\"}},\"id\":\"1159\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1126\",\"type\":\"DataRange1d\"},{\"attributes\":{\"axis_label\":\"Correlation\",\"formatter\":{\"id\":\"1174\"},\"ticker\":{\"id\":\"1135\"}},\"id\":\"1134\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"1179\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1152\",\"type\":\"Scatter\"},{\"attributes\":{},\"id\":\"1119\",\"type\":\"PanTool\"},{\"attributes\":{\"line_alpha\":0.1,\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1157\",\"type\":\"Line\"},{\"attributes\":{},\"id\":\"1178\",\"type\":\"Selection\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"1142\",\"type\":\"BoxAnnotation\"},{\"attributes\":{},\"id\":\"1139\",\"type\":\"BasicTicker\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"1117\"},{\"id\":\"1118\"},{\"id\":\"1119\"},{\"id\":\"1120\"},{\"id\":\"1121\"},{\"id\":\"1122\"}]},\"id\":\"1143\",\"type\":\"Toolbar\"},{\"attributes\":{\"overlay\":{\"id\":\"1142\"}},\"id\":\"1120\",\"type\":\"BoxZoomTool\"},{\"attributes\":{\"data\":{\"x\":{\"__ndarray__\":\"C9tn9G4N4r+hiDDyPt/ePw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]},\"y\":{\"__ndarray__\":\"C9tn9G4N4r+hiDDyPt/ePw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]}},\"selected\":{\"id\":\"1178\"},\"selection_policy\":{\"id\":\"1179\"}},\"id\":\"1155\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"axis_label\":\"Partial Correlation\",\"formatter\":{\"id\":\"1172\"},\"ticker\":{\"id\":\"1139\"}},\"id\":\"1138\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"1132\",\"type\":\"LinearScale\"},{\"attributes\":{\"data_source\":{\"id\":\"1155\"},\"glyph\":{\"id\":\"1156\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1157\"},\"selection_glyph\":null,\"view\":{\"id\":\"1159\"}},\"id\":\"1158\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"1135\",\"type\":\"BasicTicker\"},{\"attributes\":{\"data_source\":{\"id\":\"1116\"},\"glyph\":{\"id\":\"1151\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1152\"},\"selection_glyph\":null,\"view\":{\"id\":\"1154\"}},\"id\":\"1153\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"data\":{\"corr\":{\"__ndarray__\":\"IF1P1gIm3D9TkVCCIgPXPxOixxbsCtW/kbeAuno3zT9x85hPS7OTP0vlzKlX2YE/2HrfNr141T9CwphYxFXLP+ZuMw0bwdE/49961bIXxr9sJcgM513GP7bbogleOsS/v8ximaUctj8ECFRzHQXLP3U1ja71Tsg/9yLq5OkS1T8L6w8ThGiUPzysRQz+MNY/HULNg9ykt7+BFQsJoKTJPw3Poz9GOse/tbKTxGxrxT9hORHX/ZG3P2qhqDJOrdY/POZVHk+d1r87PsCG9QfVPxCPycDK99o/hre6KcnW0b+9rTcd2q6Dvwwn1sVS8rG/putdh/pjwD8J5P7cwGnUvwsV3GKFp9E/oYgw8j7f3j8qr9ciqZLJv1N6Vir7kpu/2bbzunkaqb9XPh48XHDbP4cVRhciNco/kSLiKUYJ3L/GXgVpeNyRv1tWFtO+t6I/e9G1GzSJ0L8hEReeh9GyPx4EyshZ89o/NIf1+vI7zz9qlsw3y5vBv5ByTEFgo7e/H4imfw1qo78sXeBdI7zIv9iT37yLbdY/49oc2enhxj8y4MylH5ClvwvbZ/RuDeK/xW6JJW7anz8=\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[55]},\"id\":[[\"Bmpr1a\",\"Bmpr1b\"],[\"Bmpr1a\",\"Acvr1\"],[\"Bmpr1a\",\"Acvrl1\"],[\"Bmpr1a\",\"Acvr1b\"],[\"Bmpr1a\",\"Tgfbr1\"],[\"Bmpr1a\",\"Acvr1c\"],[\"Bmpr1a\",\"Acvr2a\"],[\"Bmpr1a\",\"Acvr2b\"],[\"Bmpr1a\",\"Bmpr2\"],[\"Bmpr1a\",\"Tgfbr2\"],[\"Bmpr1b\",\"Acvr1\"],[\"Bmpr1b\",\"Acvrl1\"],[\"Bmpr1b\",\"Acvr1b\"],[\"Bmpr1b\",\"Tgfbr1\"],[\"Bmpr1b\",\"Acvr1c\"],[\"Bmpr1b\",\"Acvr2a\"],[\"Bmpr1b\",\"Acvr2b\"],[\"Bmpr1b\",\"Bmpr2\"],[\"Bmpr1b\",\"Tgfbr2\"],[\"Acvr1\",\"Acvrl1\"],[\"Acvr1\",\"Acvr1b\"],[\"Acvr1\",\"Tgfbr1\"],[\"Acvr1\",\"Acvr1c\"],[\"Acvr1\",\"Acvr2a\"],[\"Acvr1\",\"Acvr2b\"],[\"Acvr1\",\"Bmpr2\"],[\"Acvr1\",\"Tgfbr2\"],[\"Acvrl1\",\"Acvr1b\"],[\"Acvrl1\",\"Tgfbr1\"],[\"Acvrl1\",\"Acvr1c\"],[\"Acvrl1\",\"Acvr2a\"],[\"Acvrl1\",\"Acvr2b\"],[\"Acvrl1\",\"Bmpr2\"],[\"Acvrl1\",\"Tgfbr2\"],[\"Acvr1b\",\"Tgfbr1\"],[\"Acvr1b\",\"Acvr1c\"],[\"Acvr1b\",\"Acvr2a\"],[\"Acvr1b\",\"Acvr2b\"],[\"Acvr1b\",\"Bmpr2\"],[\"Acvr1b\",\"Tgfbr2\"],[\"Tgfbr1\",\"Acvr1c\"],[\"Tgfbr1\",\"Acvr2a\"],[\"Tgfbr1\",\"Acvr2b\"],[\"Tgfbr1\",\"Bmpr2\"],[\"Tgfbr1\",\"Tgfbr2\"],[\"Acvr1c\",\"Acvr2a\"],[\"Acvr1c\",\"Acvr2b\"],[\"Acvr1c\",\"Bmpr2\"],[\"Acvr1c\",\"Tgfbr2\"],[\"Acvr2a\",\"Acvr2b\"],[\"Acvr2a\",\"Bmpr2\"],[\"Acvr2a\",\"Tgfbr2\"],[\"Acvr2b\",\"Bmpr2\"],[\"Acvr2b\",\"Tgfbr2\"],[\"Bmpr2\",\"Tgfbr2\"]],\"index\":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],\"pcorr\":{\"__ndarray__\":\"c0lR4DxR0T8Rt7p8r2TcP8iVUjIY69G/AxS42lfSsz+0dmzxuip7v51htVrxvsK/Pn7Sf8AozT8avH72yBnNP6iwcMJfN6k/lW1M9ZpKt78xudG474O1v7k9JTXy2ba/TFCGD3egs7/lPp7P/y/MP9P36n7M1Mg/RaZSj/wcwD86TwAv78Glv3fMwzkUldA/9vCgNRhaub94jhowTiKYP0Oz0TnFAri/3HAg87jnob/fP95SQvbBP04Q+5T2Apg/cFphKO+7xb+r7MNCqUnPP8Fo0nPvOtM/RSYAZafmxb97zfiod5jOv6twJ4ylVI2/DIj11IF5kj+zjNjNqKyYPy/YHWwTTdo/baljlhD31j8EnRqgeaayv779uV3z4rA/jbYoy7RDpb+HUXRBTZjPP9G0rFTiBtQ/g1g1SJ3st79ZfsxXjnSIv7I9R1RixMC/1vXblkLef7+csCY3ZkrBP5KQJYyN69o/0UG96+Vr0D99cR/NAZq/v5h49yEyoMy/sECK9JpHwr/E9f9RsLCsv3n0vUy5Y84/iGFzC3Myxz8K4Mlx9SOhvwZNXj5IBdK/tewAB0aHyL8=\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[55]}},\"selected\":{\"id\":\"1176\"},\"selection_policy\":{\"id\":\"1177\"}},\"id\":\"1116\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"1172\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"source\":{\"id\":\"1116\"}},\"id\":\"1154\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1176\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1174\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{},\"id\":\"1177\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Genes\",\"@id\"]]},\"id\":\"1117\",\"type\":\"HoverTool\"}],\"root_ids\":[\"1123\"]},\"title\":\"Bokeh Application\",\"version\":\"2.1.1\"}};\n",
       "  var render_items = [{\"docid\":\"159619af-7bd9-4928-9df1-b40a1f61eb47\",\"root_ids\":[\"1123\"],\"roots\":{\"1123\":\"77c4c74f-5278-49fa-a5b7-d846ac2d35ab\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "1123"
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'/Users/niveditakanrar/Documents/SURF/Code/alejo_metadata/figures/own_norm/bmpr_int_corr_pcorr_scatter.html'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = ColumnDataSource(df)\n",
    "output_file(filename=figdir+\"bmpr_int_corr_pcorr_scatter.html\")\n",
    "\n",
    "hover = HoverTool(tooltips=[('Genes', '@id')])\n",
    "tools = [hover, WheelZoomTool(), PanTool(), BoxZoomTool(), ResetTool(), SaveTool()]\n",
    "\n",
    "p = figure(width=400, height=400,tools=tools,\n",
    "           toolbar_location=\"below\", \n",
    "           x_axis_label = \"Correlation\",\n",
    "           y_axis_label = \"Partial Correlation\",\n",
    "           title = \"BMP Receptors\")\n",
    "\n",
    "p.scatter(x = \"corr\", y = \"pcorr\", source = source, color = \"#1f77b4\");\n",
    "p.line(min_max[:,0], min_max[:,1], line_color = \"black\");\n",
    "\n",
    "show(p)\n",
    "save(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Wnt Receptors<a name=\"wntr\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will use the list of Notch components stored in our curated list above. Otherwise, if you want to study a set of signaling genes not in this csv, simply change the pathway name and list of genes below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the Notch genes from our reference list\n",
    "pathway = \"Wnt\"\n",
    "genes = list(pathways[pathways[\"pathway\"] == pathway][\"gene\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As of now, the integrated atlas only consists of the intersection of genes that are expressed at a basal level in all datasets (for a total of 11,675 genes out of ~22k total). Some genes specified in our list above may not be contained within this intersection, so we filter them out below, and subset the remaining pathway genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fzd1</th>\n",
       "      <th>Fzd2</th>\n",
       "      <th>Fzd3</th>\n",
       "      <th>Fzd4</th>\n",
       "      <th>Fzd5</th>\n",
       "      <th>Fzd6</th>\n",
       "      <th>Fzd7</th>\n",
       "      <th>Fzd8</th>\n",
       "      <th>Fzd9</th>\n",
       "      <th>Fzd10</th>\n",
       "      <th>Lrp5</th>\n",
       "      <th>Lrp6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.035543</td>\n",
       "      <td>0.402745</td>\n",
       "      <td>0.094570</td>\n",
       "      <td>0.005178</td>\n",
       "      <td>0.055711</td>\n",
       "      <td>0.014275</td>\n",
       "      <td>0.252216</td>\n",
       "      <td>0.023789</td>\n",
       "      <td>0.024779</td>\n",
       "      <td>0.063997</td>\n",
       "      <td>0.114316</td>\n",
       "      <td>0.296308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127067</td>\n",
       "      <td>0.050725</td>\n",
       "      <td>0.008090</td>\n",
       "      <td>0.006428</td>\n",
       "      <td>0.006039</td>\n",
       "      <td>0.051071</td>\n",
       "      <td>0.003112</td>\n",
       "      <td>0.005319</td>\n",
       "      <td>0.010004</td>\n",
       "      <td>0.030376</td>\n",
       "      <td>0.027998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.014320</td>\n",
       "      <td>0.450853</td>\n",
       "      <td>0.062696</td>\n",
       "      <td>0.020652</td>\n",
       "      <td>0.048098</td>\n",
       "      <td>0.005842</td>\n",
       "      <td>0.335859</td>\n",
       "      <td>0.021437</td>\n",
       "      <td>0.012596</td>\n",
       "      <td>0.027045</td>\n",
       "      <td>0.150844</td>\n",
       "      <td>0.286286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.016819</td>\n",
       "      <td>0.462213</td>\n",
       "      <td>0.105133</td>\n",
       "      <td>0.013894</td>\n",
       "      <td>0.082077</td>\n",
       "      <td>0.029061</td>\n",
       "      <td>0.236436</td>\n",
       "      <td>0.023649</td>\n",
       "      <td>0.012668</td>\n",
       "      <td>0.001679</td>\n",
       "      <td>0.119725</td>\n",
       "      <td>0.253350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.100684</td>\n",
       "      <td>0.032882</td>\n",
       "      <td>0.051082</td>\n",
       "      <td>0.009923</td>\n",
       "      <td>0.024890</td>\n",
       "      <td>0.007970</td>\n",
       "      <td>0.111538</td>\n",
       "      <td>0.010448</td>\n",
       "      <td>0.020089</td>\n",
       "      <td>0.013549</td>\n",
       "      <td>0.109442</td>\n",
       "      <td>0.212099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Fzd1      Fzd2      Fzd3      Fzd4      Fzd5      Fzd6  \\\n",
       "1_E6.5_8.5_Chan  0.035543  0.402745  0.094570  0.005178  0.055711  0.014275   \n",
       "2_E6.5_8.5_Chan  0.000000  0.127067  0.050725  0.008090  0.006428  0.006039   \n",
       "3_E6.5_8.5_Chan  0.014320  0.450853  0.062696  0.020652  0.048098  0.005842   \n",
       "4_E6.5_8.5_Chan  0.016819  0.462213  0.105133  0.013894  0.082077  0.029061   \n",
       "5_E6.5_8.5_Chan  0.100684  0.032882  0.051082  0.009923  0.024890  0.007970   \n",
       "\n",
       "                     Fzd7      Fzd8      Fzd9     Fzd10      Lrp5      Lrp6  \n",
       "1_E6.5_8.5_Chan  0.252216  0.023789  0.024779  0.063997  0.114316  0.296308  \n",
       "2_E6.5_8.5_Chan  0.051071  0.003112  0.005319  0.010004  0.030376  0.027998  \n",
       "3_E6.5_8.5_Chan  0.335859  0.021437  0.012596  0.027045  0.150844  0.286286  \n",
       "4_E6.5_8.5_Chan  0.236436  0.023649  0.012668  0.001679  0.119725  0.253350  \n",
       "5_E6.5_8.5_Chan  0.111538  0.010448  0.020089  0.013549  0.109442  0.212099  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset a DataFrame with only the genes selected above\n",
    "pathway_genes = [i for i in genes if i in integrated_counts.columns]\n",
    "\n",
    "df_pathway = integrated_counts[pathway_genes]\n",
    "\n",
    "df_pathway.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Scaling of Counts <a name=\"scaling\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compare expression of different genes in a pathway, we need to account for the the fact that counts for different genes may have very different dynamic ranges, and any direct clustering on these normalized data would be biased toward highly-expressed genes. To give equal weight to all features, we can perform MinMax scaling to scale counts per gene to lie in the range of 0 to 1. However, since RNA count distributions are right-tailed, scaling the maximum count value per gene to correspond to a value of 1 would sharply skew the new, scaled. values. We can instead more robustly saturate count values by scaling the count value that corresponds to a scaled expression value of 1.0 to be some percentile of maximum gene expression (i.e. the 99th percentile). \n",
    "\n",
    "The function below saturates the expression values in a DataFrame at inputted gene-specific percentiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fzd1</th>\n",
       "      <th>Fzd2</th>\n",
       "      <th>Fzd3</th>\n",
       "      <th>Fzd4</th>\n",
       "      <th>Fzd5</th>\n",
       "      <th>Fzd6</th>\n",
       "      <th>Fzd7</th>\n",
       "      <th>Fzd8</th>\n",
       "      <th>Fzd9</th>\n",
       "      <th>Fzd10</th>\n",
       "      <th>Lrp5</th>\n",
       "      <th>Lrp6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.072044</td>\n",
       "      <td>0.147088</td>\n",
       "      <td>0.063665</td>\n",
       "      <td>0.054830</td>\n",
       "      <td>0.043207</td>\n",
       "      <td>0.044121</td>\n",
       "      <td>0.068572</td>\n",
       "      <td>0.032015</td>\n",
       "      <td>0.012256</td>\n",
       "      <td>0.028927</td>\n",
       "      <td>0.084144</td>\n",
       "      <td>0.176863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.131776</td>\n",
       "      <td>0.226803</td>\n",
       "      <td>0.097499</td>\n",
       "      <td>0.106357</td>\n",
       "      <td>0.064917</td>\n",
       "      <td>0.091257</td>\n",
       "      <td>0.086896</td>\n",
       "      <td>0.048795</td>\n",
       "      <td>0.033232</td>\n",
       "      <td>0.070483</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>0.144163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.002275</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000765</td>\n",
       "      <td>0.001006</td>\n",
       "      <td>0.005678</td>\n",
       "      <td>0.000994</td>\n",
       "      <td>0.006335</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017946</td>\n",
       "      <td>0.066768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.026826</td>\n",
       "      <td>0.014881</td>\n",
       "      <td>0.025475</td>\n",
       "      <td>0.010813</td>\n",
       "      <td>0.016631</td>\n",
       "      <td>0.007595</td>\n",
       "      <td>0.032716</td>\n",
       "      <td>0.010705</td>\n",
       "      <td>0.001694</td>\n",
       "      <td>0.000569</td>\n",
       "      <td>0.051630</td>\n",
       "      <td>0.134056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.079249</td>\n",
       "      <td>0.231676</td>\n",
       "      <td>0.087851</td>\n",
       "      <td>0.051063</td>\n",
       "      <td>0.052708</td>\n",
       "      <td>0.030877</td>\n",
       "      <td>0.100439</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.009486</td>\n",
       "      <td>0.019242</td>\n",
       "      <td>0.104989</td>\n",
       "      <td>0.249414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.849332</td>\n",
       "      <td>0.864560</td>\n",
       "      <td>0.552073</td>\n",
       "      <td>0.559360</td>\n",
       "      <td>0.350935</td>\n",
       "      <td>0.471960</td>\n",
       "      <td>0.412889</td>\n",
       "      <td>0.236950</td>\n",
       "      <td>0.240949</td>\n",
       "      <td>0.419822</td>\n",
       "      <td>0.485859</td>\n",
       "      <td>0.688064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Fzd1         Fzd2         Fzd3         Fzd4         Fzd5  \\\n",
       "count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   \n",
       "mean      0.072044     0.147088     0.063665     0.054830     0.043207   \n",
       "std       0.131776     0.226803     0.097499     0.106357     0.064917   \n",
       "min       0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "25%       0.002275     0.000000     0.000765     0.001006     0.005678   \n",
       "50%       0.026826     0.014881     0.025475     0.010813     0.016631   \n",
       "75%       0.079249     0.231676     0.087851     0.051063     0.052708   \n",
       "max       0.849332     0.864560     0.552073     0.559360     0.350935   \n",
       "\n",
       "              Fzd6         Fzd7         Fzd8         Fzd9        Fzd10  \\\n",
       "count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   \n",
       "mean      0.044121     0.068572     0.032015     0.012256     0.028927   \n",
       "std       0.091257     0.086896     0.048795     0.033232     0.070483   \n",
       "min       0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "25%       0.000994     0.006335     0.000000     0.000000     0.000000   \n",
       "50%       0.007595     0.032716     0.010705     0.001694     0.000569   \n",
       "75%       0.030877     0.100439     0.039046     0.009486     0.019242   \n",
       "max       0.471960     0.412889     0.236950     0.240949     0.419822   \n",
       "\n",
       "              Lrp5         Lrp6  \n",
       "count  1206.000000  1206.000000  \n",
       "mean      0.084144     0.176863  \n",
       "std       0.099900     0.144163  \n",
       "min       0.000000     0.000000  \n",
       "25%       0.017946     0.066768  \n",
       "50%       0.051630     0.134056  \n",
       "75%       0.104989     0.249414  \n",
       "max       0.485859     0.688064  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute the 99th quantile per gene\n",
    "gene_quantiles = np.quantile(df_pathway, q = 0.99, axis = 0)\n",
    "\n",
    "# Clip the values such that any counts above the 99th quantile take its value\n",
    "x_clipped = quantile_saturation(df_pathway, gene_quantiles)\n",
    "x_clipped.describe() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now scale expression values in this DataFrame of \"clipped\" counts to lie between 0 and 1, using a simple MinMaxScaler from scikitlearn that linearly scales counts to lie between the a specified minimum and maximum value per gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Fzd1</th>\n",
       "      <th>Fzd2</th>\n",
       "      <th>Fzd3</th>\n",
       "      <th>Fzd4</th>\n",
       "      <th>Fzd5</th>\n",
       "      <th>Fzd6</th>\n",
       "      <th>Fzd7</th>\n",
       "      <th>Fzd8</th>\n",
       "      <th>Fzd9</th>\n",
       "      <th>Fzd10</th>\n",
       "      <th>Lrp5</th>\n",
       "      <th>Lrp6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.041848</td>\n",
       "      <td>0.465838</td>\n",
       "      <td>0.171300</td>\n",
       "      <td>0.009257</td>\n",
       "      <td>0.158750</td>\n",
       "      <td>0.030246</td>\n",
       "      <td>0.610857</td>\n",
       "      <td>0.100397</td>\n",
       "      <td>0.102840</td>\n",
       "      <td>0.152438</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.430640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.146973</td>\n",
       "      <td>0.091882</td>\n",
       "      <td>0.014462</td>\n",
       "      <td>0.018317</td>\n",
       "      <td>0.012796</td>\n",
       "      <td>0.123691</td>\n",
       "      <td>0.013136</td>\n",
       "      <td>0.022075</td>\n",
       "      <td>0.023829</td>\n",
       "      <td>0.062521</td>\n",
       "      <td>0.040691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.016860</td>\n",
       "      <td>0.521482</td>\n",
       "      <td>0.113565</td>\n",
       "      <td>0.036922</td>\n",
       "      <td>0.137058</td>\n",
       "      <td>0.012377</td>\n",
       "      <td>0.813435</td>\n",
       "      <td>0.090470</td>\n",
       "      <td>0.052278</td>\n",
       "      <td>0.064421</td>\n",
       "      <td>0.310469</td>\n",
       "      <td>0.416074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.019803</td>\n",
       "      <td>0.534622</td>\n",
       "      <td>0.190433</td>\n",
       "      <td>0.024840</td>\n",
       "      <td>0.233882</td>\n",
       "      <td>0.061575</td>\n",
       "      <td>0.572637</td>\n",
       "      <td>0.099804</td>\n",
       "      <td>0.052574</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.246419</td>\n",
       "      <td>0.368207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.118545</td>\n",
       "      <td>0.038033</td>\n",
       "      <td>0.092527</td>\n",
       "      <td>0.017741</td>\n",
       "      <td>0.070925</td>\n",
       "      <td>0.016887</td>\n",
       "      <td>0.270140</td>\n",
       "      <td>0.044095</td>\n",
       "      <td>0.083374</td>\n",
       "      <td>0.032273</td>\n",
       "      <td>0.225256</td>\n",
       "      <td>0.308255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Fzd1      Fzd2      Fzd3      Fzd4      Fzd5      Fzd6  \\\n",
       "1_E6.5_8.5_Chan  0.041848  0.465838  0.171300  0.009257  0.158750  0.030246   \n",
       "2_E6.5_8.5_Chan  0.000000  0.146973  0.091882  0.014462  0.018317  0.012796   \n",
       "3_E6.5_8.5_Chan  0.016860  0.521482  0.113565  0.036922  0.137058  0.012377   \n",
       "4_E6.5_8.5_Chan  0.019803  0.534622  0.190433  0.024840  0.233882  0.061575   \n",
       "5_E6.5_8.5_Chan  0.118545  0.038033  0.092527  0.017741  0.070925  0.016887   \n",
       "\n",
       "                     Fzd7      Fzd8      Fzd9     Fzd10      Lrp5      Lrp6  \n",
       "1_E6.5_8.5_Chan  0.610857  0.100397  0.102840  0.152438  0.235286  0.430640  \n",
       "2_E6.5_8.5_Chan  0.123691  0.013136  0.022075  0.023829  0.062521  0.040691  \n",
       "3_E6.5_8.5_Chan  0.813435  0.090470  0.052278  0.064421  0.310469  0.416074  \n",
       "4_E6.5_8.5_Chan  0.572637  0.099804  0.052574  0.004000  0.246419  0.368207  \n",
       "5_E6.5_8.5_Chan  0.270140  0.044095  0.083374  0.032273  0.225256  0.308255  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize the scaler\n",
    "scaler = MinMaxScaler()\n",
    "# Fit the scaler to the dataset to compute per-feature min. and max. values\n",
    "scaler.fit(x_clipped)\n",
    "\n",
    "# Apply the scaler to the dataset\n",
    "df_scale = pd.DataFrame(scaler.transform(x_clipped), columns = x_clipped.columns, index = x_clipped.index)\n",
    "\n",
    "df_scale.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### WntR Correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at gene correlation for clusters that have the BMP receptors active. We can try to measure this by first computing correlation only at clusters that have expression of 2 receptors greater than or equal to 0.2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(813, 12)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filt = pd.DataFrame(columns = df_scale.columns)\n",
    "\n",
    "for i in df_scale.index:\n",
    "    if (df_scale.loc[i][pathway_genes] >= 0.2).sum() > 0:\n",
    "        df_filt = df_filt.append(df_scale.loc[i])\n",
    "        \n",
    "df_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_corr = as.data.frame(cor(df_filt))\n",
    "colnames(M_corr) = colnames(df_filt)\n",
    "rownames(M_corr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/wntr_corrplot_all.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_corr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### WntR Partial Correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_pcorr = as.data.frame(pcor(df_filt)$estimate)\n",
    "colnames(M_pcorr) = colnames(df_filt)\n",
    "rownames(M_pcorr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/wntr_parcorrplot_all.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_pcorr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "%R -o M_corr,M_pcorr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_corr = M_corr.where(np.triu(np.ones(M_corr.shape), k = 1).astype(np.bool))\n",
    "M_pcorr = M_pcorr.where(np.triu(np.ones(M_pcorr.shape), k = 1).astype(np.bool))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_max = np.zeros((2,2))\n",
    "\n",
    "min_max[0] = min(np.nanmin(M_corr.min(skipna = True)), np.nanmin(M_pcorr.min(skipna = True)))\n",
    "min_max[1] = max(np.nanmax(M_corr.max(skipna = True)), np.nanmax(M_pcorr.max(skipna = True)))\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for i in list(itertools.combinations(df_filt.columns, 2)):\n",
    "    x = pd.DataFrame()\n",
    "    df = df.append([[M_corr.loc[i[0]][i[1]], M_pcorr.loc[i[0]][i[1]], i]])\n",
    "    \n",
    "df.columns = [\"corr\", \"pcorr\", \"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"96113f55-ea3d-428b-9198-95ade72156c4\" data-root-id=\"1509\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"0b41962f-f78f-4204-9df6-ed5b83fd734b\":{\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"1520\"}],\"center\":[{\"id\":\"1523\"},{\"id\":\"1527\"}],\"left\":[{\"id\":\"1524\"}],\"plot_height\":400,\"plot_width\":400,\"renderers\":[{\"id\":\"1539\"},{\"id\":\"1544\"}],\"title\":{\"id\":\"1510\"},\"toolbar\":{\"id\":\"1529\"},\"toolbar_location\":\"below\",\"x_range\":{\"id\":\"1512\"},\"x_scale\":{\"id\":\"1516\"},\"y_range\":{\"id\":\"1514\"},\"y_scale\":{\"id\":\"1518\"}},\"id\":\"1509\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"data_source\":{\"id\":\"1541\"},\"glyph\":{\"id\":\"1542\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1543\"},\"selection_glyph\":null,\"view\":{\"id\":\"1545\"}},\"id\":\"1544\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Genes\",\"@id\"]]},\"id\":\"1503\",\"type\":\"HoverTool\"},{\"attributes\":{},\"id\":\"1525\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"1518\",\"type\":\"LinearScale\"},{\"attributes\":{},\"id\":\"1570\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"axis_label\":\"Correlation\",\"formatter\":{\"id\":\"1570\"},\"ticker\":{\"id\":\"1521\"}},\"id\":\"1520\",\"type\":\"LinearAxis\"},{\"attributes\":{\"axis_label\":\"Partial Correlation\",\"formatter\":{\"id\":\"1568\"},\"ticker\":{\"id\":\"1525\"}},\"id\":\"1524\",\"type\":\"LinearAxis\"},{\"attributes\":{\"data_source\":{\"id\":\"1502\"},\"glyph\":{\"id\":\"1537\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1538\"},\"selection_glyph\":null,\"view\":{\"id\":\"1540\"}},\"id\":\"1539\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"line_alpha\":0.1,\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1543\",\"type\":\"Line\"},{\"attributes\":{},\"id\":\"1568\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{},\"id\":\"1521\",\"type\":\"BasicTicker\"},{\"attributes\":{\"axis\":{\"id\":\"1524\"},\"dimension\":1,\"ticker\":null},\"id\":\"1527\",\"type\":\"Grid\"},{\"attributes\":{\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1542\",\"type\":\"Line\"},{\"attributes\":{},\"id\":\"1574\",\"type\":\"Selection\"},{\"attributes\":{\"axis\":{\"id\":\"1520\"},\"ticker\":null},\"id\":\"1523\",\"type\":\"Grid\"},{\"attributes\":{\"text\":\"Wnt Receptors\"},\"id\":\"1510\",\"type\":\"Title\"},{\"attributes\":{\"data\":{\"x\":{\"__ndarray__\":\"s9s/gRlU0b8t6yfCzzLjPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]},\"y\":{\"__ndarray__\":\"s9s/gRlU0b8t6yfCzzLjPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]}},\"selected\":{\"id\":\"1574\"},\"selection_policy\":{\"id\":\"1575\"}},\"id\":\"1541\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"1508\",\"type\":\"SaveTool\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"1503\"},{\"id\":\"1504\"},{\"id\":\"1505\"},{\"id\":\"1506\"},{\"id\":\"1507\"},{\"id\":\"1508\"}]},\"id\":\"1529\",\"type\":\"Toolbar\"},{\"attributes\":{},\"id\":\"1514\",\"type\":\"DataRange1d\"},{\"attributes\":{},\"id\":\"1572\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1573\",\"type\":\"UnionRenderers\"},{\"attributes\":{},\"id\":\"1575\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1538\",\"type\":\"Scatter\"},{\"attributes\":{\"source\":{\"id\":\"1502\"}},\"id\":\"1540\",\"type\":\"CDSView\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"1528\",\"type\":\"BoxAnnotation\"},{\"attributes\":{},\"id\":\"1504\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"1507\",\"type\":\"ResetTool\"},{\"attributes\":{},\"id\":\"1512\",\"type\":\"DataRange1d\"},{\"attributes\":{\"data\":{\"corr\":{\"__ndarray__\":\"3IxytMsaxj/iIaRWXiK0P5hmkg0V6H6/VrenKYnImz/q/WJo1Tx/v12+AvfXAdY/ABGuLudm0T9y+tHc74GgP73rqXTwz5w/6VUeAjx9tj9uRqncKx/RP9TBjMuBr9Y/Yo440Z5uyL9n0a79DAHOv1i5AGXs08+/5muscYB43j+YCTYX3ZWxP4taWZEfQ8k/x4dQIZFD1D+z2z+BGVTRv8NeN71CwIE/VQQBajh3wr/Ct84YAq+/v8iXJgQloX2/TqjatS4GxT8nT9hcSkWMv5f8vgzAW8Q/zEm6WdxqxT+rhXMIwVLKv6FJ2pvar7A/qj5QG1fhwz8OrUQxiArhPx1uxEvGcKq/NyBfoK0I1z/Ivcb+5f3Av9OGq8OS0ci/se8zkgW83z/xTDu0ksjTPz2sQCvh980/hBYtpjkewz8YlY77HzjcPy5DGcgF36O/z1a0JDsaz7+R8Gp9vGXTP5nCpTNoOto/Swb/qu1Yq78u/MAcu3nUP2hEfpNRNLS/8VeK4pfpoD8t6yfCzzLjP9c3o0YS8dA/XYCe3FaSzD8gr3o+7U+nP+/T88pq3IK/PgMnsrEqvz8i5Fx/0OfTP544O1b9oLs/ULreQvAmwb+0iRwLgxTYP6547aml894/MLkpMbyAuz/o5ffK+dGlv6Tfl9J7Yag/qty/dxdPvL+MN5ty866iv9tXC+SPRtk/\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[66]},\"id\":[[\"Fzd1\",\"Fzd2\"],[\"Fzd1\",\"Fzd3\"],[\"Fzd1\",\"Fzd4\"],[\"Fzd1\",\"Fzd5\"],[\"Fzd1\",\"Fzd6\"],[\"Fzd1\",\"Fzd7\"],[\"Fzd1\",\"Fzd8\"],[\"Fzd1\",\"Fzd9\"],[\"Fzd1\",\"Fzd10\"],[\"Fzd1\",\"Lrp5\"],[\"Fzd1\",\"Lrp6\"],[\"Fzd2\",\"Fzd3\"],[\"Fzd2\",\"Fzd4\"],[\"Fzd2\",\"Fzd5\"],[\"Fzd2\",\"Fzd6\"],[\"Fzd2\",\"Fzd7\"],[\"Fzd2\",\"Fzd8\"],[\"Fzd2\",\"Fzd9\"],[\"Fzd2\",\"Fzd10\"],[\"Fzd2\",\"Lrp5\"],[\"Fzd2\",\"Lrp6\"],[\"Fzd3\",\"Fzd4\"],[\"Fzd3\",\"Fzd5\"],[\"Fzd3\",\"Fzd6\"],[\"Fzd3\",\"Fzd7\"],[\"Fzd3\",\"Fzd8\"],[\"Fzd3\",\"Fzd9\"],[\"Fzd3\",\"Fzd10\"],[\"Fzd3\",\"Lrp5\"],[\"Fzd3\",\"Lrp6\"],[\"Fzd4\",\"Fzd5\"],[\"Fzd4\",\"Fzd6\"],[\"Fzd4\",\"Fzd7\"],[\"Fzd4\",\"Fzd8\"],[\"Fzd4\",\"Fzd9\"],[\"Fzd4\",\"Fzd10\"],[\"Fzd4\",\"Lrp5\"],[\"Fzd4\",\"Lrp6\"],[\"Fzd5\",\"Fzd6\"],[\"Fzd5\",\"Fzd7\"],[\"Fzd5\",\"Fzd8\"],[\"Fzd5\",\"Fzd9\"],[\"Fzd5\",\"Fzd10\"],[\"Fzd5\",\"Lrp5\"],[\"Fzd5\",\"Lrp6\"],[\"Fzd6\",\"Fzd7\"],[\"Fzd6\",\"Fzd8\"],[\"Fzd6\",\"Fzd9\"],[\"Fzd6\",\"Fzd10\"],[\"Fzd6\",\"Lrp5\"],[\"Fzd6\",\"Lrp6\"],[\"Fzd7\",\"Fzd8\"],[\"Fzd7\",\"Fzd9\"],[\"Fzd7\",\"Fzd10\"],[\"Fzd7\",\"Lrp5\"],[\"Fzd7\",\"Lrp6\"],[\"Fzd8\",\"Fzd9\"],[\"Fzd8\",\"Fzd10\"],[\"Fzd8\",\"Lrp5\"],[\"Fzd8\",\"Lrp6\"],[\"Fzd9\",\"Fzd10\"],[\"Fzd9\",\"Lrp5\"],[\"Fzd9\",\"Lrp6\"],[\"Fzd10\",\"Lrp5\"],[\"Fzd10\",\"Lrp6\"],[\"Lrp5\",\"Lrp6\"]],\"index\":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],\"pcorr\":{\"__ndarray__\":\"wpWbpViHp7/ZI+kd4vSHP+Qqk2U+WbW/3LTXZ+3OxL/Hgxg9EWClvwq3yipwwM0/EziGWdPFzD8jzFr7yEujv2zWAkGxUqA/73x3Oc50gT/ap68JYKjCP4ZmUVGAJ8w//a/i/XRBsz8mFy2T4cHNv/XmhVtr0cK/TEh9nJOh4D/ufh8fzqPNPzq6DriLD70/kGK1dpK/1D8fKbggvq7Jv5JEpFCIdLG/ffDDuGEMur/4/5l8HUqxv/N6Ri12Dc0/WOtHhQVxnT9p2c6BFqmlv77ccc6cbrk/qkRCjCa5i79LZjaAk7XKv10m0DXiicI/6Or5orvVxb+PbIQgMqTVP1OXwsbl4L+/EGfXqU9vxz9V00BCmPzAv8lYcX1toMy/ks2Ydx8yxj9ARflBuyvHP6NIrgwllLU/rHXsHVTWxD8OQuoVlaPVP9bkQDtKnaW/tPAT7tIewb9/N/cbkeGAv5yRPh4hLdA/U9uPBXeEqr8WT1c8QRm8P0boHGI3R6y/4wnyxj7Xzj86UBYDcCnbPwTNzZJ5TLG/CqRjvOX0r7+hkllGCUW1v8J2a8NtPsa/p/21imL8zD8Hg0p9ytLGP6ZilzDwbMI/y9BUFcnFvb9rbMA1w0e1P9QhQH/Odck/KRL9HOFroD8CL0P4efOxP9wWkKkWh6Q/K2FwB0RRgj9AgHowYAu/PyG6hgJQm8Q/\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[66]}},\"selected\":{\"id\":\"1572\"},\"selection_policy\":{\"id\":\"1573\"}},\"id\":\"1502\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"1516\",\"type\":\"LinearScale\"},{\"attributes\":{\"overlay\":{\"id\":\"1528\"}},\"id\":\"1506\",\"type\":\"BoxZoomTool\"},{\"attributes\":{\"fill_color\":{\"value\":\"#1f77b4\"},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1537\",\"type\":\"Scatter\"},{\"attributes\":{\"source\":{\"id\":\"1541\"}},\"id\":\"1545\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1505\",\"type\":\"PanTool\"}],\"root_ids\":[\"1509\"]},\"title\":\"Bokeh Application\",\"version\":\"2.1.1\"}};\n",
       "  var render_items = [{\"docid\":\"0b41962f-f78f-4204-9df6-ed5b83fd734b\",\"root_ids\":[\"1509\"],\"roots\":{\"1509\":\"96113f55-ea3d-428b-9198-95ade72156c4\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "1509"
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'/Users/niveditakanrar/Documents/SURF/Code/alejo_metadata/figures/own_norm/wntr_int_corr_pcorr_scatter.html'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = ColumnDataSource(df)\n",
    "output_file(filename=figdir+\"wntr_int_corr_pcorr_scatter.html\")\n",
    "\n",
    "hover = HoverTool(tooltips=[('Genes', '@id')])\n",
    "tools = [hover, WheelZoomTool(), PanTool(), BoxZoomTool(), ResetTool(), SaveTool()]\n",
    "\n",
    "p = figure(width=400, height=400,tools=tools,\n",
    "           toolbar_location=\"below\", \n",
    "           x_axis_label = \"Correlation\",\n",
    "           y_axis_label = \"Partial Correlation\",\n",
    "           title = \"Wnt Receptors\")\n",
    "\n",
    "p.scatter(x = \"corr\", y = \"pcorr\", source = source, color = \"#1f77b4\");\n",
    "p.line(min_max[:,0], min_max[:,1], line_color = \"black\");\n",
    "\n",
    "show(p)\n",
    "save(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notch<a name=\"notch\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will use the list of Notch components stored in our curated list above. Otherwise, if you want to study a set of signaling genes not in this csv, simply change the pathway name and list of genes below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the Notch genes from our reference list\n",
    "pathway = \"Notch\"\n",
    "genes = list(pathways[pathways[\"pathway\"] == pathway][\"gene\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As of now, the integrated atlas only consists of the intersection of genes that are expressed at a basal level in all datasets (for a total of 11,675 genes out of ~22k total). Some genes specified in our list above may not be contained within this intersection, so we filter them out below, and subset the remaining pathway genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dll1</th>\n",
       "      <th>Dll3</th>\n",
       "      <th>Dll4</th>\n",
       "      <th>Jag1</th>\n",
       "      <th>Jag2</th>\n",
       "      <th>Notch1</th>\n",
       "      <th>Notch2</th>\n",
       "      <th>Notch3</th>\n",
       "      <th>Notch4</th>\n",
       "      <th>Mfng</th>\n",
       "      <th>Rfng</th>\n",
       "      <th>Lfng</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.009960</td>\n",
       "      <td>0.128945</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.060378</td>\n",
       "      <td>0.005043</td>\n",
       "      <td>0.073015</td>\n",
       "      <td>0.081050</td>\n",
       "      <td>0.114322</td>\n",
       "      <td>0.069280</td>\n",
       "      <td>0.022600</td>\n",
       "      <td>0.081819</td>\n",
       "      <td>0.018349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.065158</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008106</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002764</td>\n",
       "      <td>0.012731</td>\n",
       "      <td>0.015500</td>\n",
       "      <td>0.043117</td>\n",
       "      <td>0.029302</td>\n",
       "      <td>0.063087</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.007375</td>\n",
       "      <td>0.069647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.059349</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056679</td>\n",
       "      <td>0.553265</td>\n",
       "      <td>0.111313</td>\n",
       "      <td>0.037587</td>\n",
       "      <td>0.054053</td>\n",
       "      <td>0.065154</td>\n",
       "      <td>0.030220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.004120</td>\n",
       "      <td>0.083725</td>\n",
       "      <td>0.002463</td>\n",
       "      <td>0.012561</td>\n",
       "      <td>0.002244</td>\n",
       "      <td>0.051420</td>\n",
       "      <td>0.041818</td>\n",
       "      <td>0.105170</td>\n",
       "      <td>0.061674</td>\n",
       "      <td>0.062460</td>\n",
       "      <td>0.082392</td>\n",
       "      <td>0.032347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.001081</td>\n",
       "      <td>0.054615</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001974</td>\n",
       "      <td>0.009948</td>\n",
       "      <td>0.024286</td>\n",
       "      <td>0.048901</td>\n",
       "      <td>0.003096</td>\n",
       "      <td>0.097439</td>\n",
       "      <td>0.006813</td>\n",
       "      <td>0.148774</td>\n",
       "      <td>0.001390</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Dll1      Dll3      Dll4      Jag1      Jag2    Notch1  \\\n",
       "1_E6.5_8.5_Chan  0.009960  0.128945  0.000000  0.060378  0.005043  0.073015   \n",
       "2_E6.5_8.5_Chan  0.000000  0.065158  0.000000  0.008106  0.000000  0.002764   \n",
       "3_E6.5_8.5_Chan  0.007375  0.069647  0.000000  0.059349  0.000000  0.056679   \n",
       "4_E6.5_8.5_Chan  0.004120  0.083725  0.002463  0.012561  0.002244  0.051420   \n",
       "5_E6.5_8.5_Chan  0.001081  0.054615  0.000000  0.001974  0.009948  0.024286   \n",
       "\n",
       "                   Notch2    Notch3    Notch4      Mfng      Rfng      Lfng  \n",
       "1_E6.5_8.5_Chan  0.081050  0.114322  0.069280  0.022600  0.081819  0.018349  \n",
       "2_E6.5_8.5_Chan  0.012731  0.015500  0.043117  0.029302  0.063087  0.000000  \n",
       "3_E6.5_8.5_Chan  0.553265  0.111313  0.037587  0.054053  0.065154  0.030220  \n",
       "4_E6.5_8.5_Chan  0.041818  0.105170  0.061674  0.062460  0.082392  0.032347  \n",
       "5_E6.5_8.5_Chan  0.048901  0.003096  0.097439  0.006813  0.148774  0.001390  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset a DataFrame with only the genes selected above\n",
    "pathway_genes = [i for i in genes if i in integrated_counts.columns]\n",
    "\n",
    "df_pathway = integrated_counts[pathway_genes]\n",
    "\n",
    "df_pathway.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Scaling of Counts <a name=\"scaling\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compare expression of different genes in a pathway, we need to account for the the fact that counts for different genes may have very different dynamic ranges, and any direct clustering on these normalized data would be biased toward highly-expressed genes. To give equal weight to all features, we can perform MinMax scaling to scale counts per gene to lie in the range of 0 to 1. However, since RNA count distributions are right-tailed, scaling the maximum count value per gene to correspond to a value of 1 would sharply skew the new, scaled. values. We can instead more robustly saturate count values by scaling the count value that corresponds to a scaled expression value of 1.0 to be some percentile of maximum gene expression (i.e. the 99th percentile). \n",
    "\n",
    "The function below saturates the expression values in a DataFrame at inputted gene-specific percentiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dll1</th>\n",
       "      <th>Dll3</th>\n",
       "      <th>Dll4</th>\n",
       "      <th>Jag1</th>\n",
       "      <th>Jag2</th>\n",
       "      <th>Notch1</th>\n",
       "      <th>Notch2</th>\n",
       "      <th>Notch3</th>\n",
       "      <th>Notch4</th>\n",
       "      <th>Mfng</th>\n",
       "      <th>Rfng</th>\n",
       "      <th>Lfng</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "      <td>1206.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.045711</td>\n",
       "      <td>0.037356</td>\n",
       "      <td>0.047563</td>\n",
       "      <td>0.132162</td>\n",
       "      <td>0.020159</td>\n",
       "      <td>0.254916</td>\n",
       "      <td>0.270746</td>\n",
       "      <td>0.085560</td>\n",
       "      <td>0.048047</td>\n",
       "      <td>0.114217</td>\n",
       "      <td>0.115788</td>\n",
       "      <td>0.165934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.105058</td>\n",
       "      <td>0.130356</td>\n",
       "      <td>0.149868</td>\n",
       "      <td>0.244890</td>\n",
       "      <td>0.050763</td>\n",
       "      <td>0.327232</td>\n",
       "      <td>0.277195</td>\n",
       "      <td>0.250395</td>\n",
       "      <td>0.109029</td>\n",
       "      <td>0.172868</td>\n",
       "      <td>0.070568</td>\n",
       "      <td>0.264093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005869</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050362</td>\n",
       "      <td>0.058762</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.003831</td>\n",
       "      <td>0.002875</td>\n",
       "      <td>0.067396</td>\n",
       "      <td>0.009333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.007271</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000660</td>\n",
       "      <td>0.033716</td>\n",
       "      <td>0.002129</td>\n",
       "      <td>0.132765</td>\n",
       "      <td>0.166685</td>\n",
       "      <td>0.014431</td>\n",
       "      <td>0.014120</td>\n",
       "      <td>0.030390</td>\n",
       "      <td>0.097068</td>\n",
       "      <td>0.040172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.036400</td>\n",
       "      <td>0.016172</td>\n",
       "      <td>0.008880</td>\n",
       "      <td>0.140159</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.301811</td>\n",
       "      <td>0.406283</td>\n",
       "      <td>0.066928</td>\n",
       "      <td>0.036206</td>\n",
       "      <td>0.178138</td>\n",
       "      <td>0.151794</td>\n",
       "      <td>0.181560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.672066</td>\n",
       "      <td>0.946207</td>\n",
       "      <td>0.822799</td>\n",
       "      <td>1.471075</td>\n",
       "      <td>0.299035</td>\n",
       "      <td>1.567202</td>\n",
       "      <td>1.287853</td>\n",
       "      <td>1.770191</td>\n",
       "      <td>0.605964</td>\n",
       "      <td>0.853848</td>\n",
       "      <td>0.346051</td>\n",
       "      <td>1.167906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Dll1         Dll3         Dll4         Jag1         Jag2  \\\n",
       "count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   \n",
       "mean      0.045711     0.037356     0.047563     0.132162     0.020159   \n",
       "std       0.105058     0.130356     0.149868     0.244890     0.050763   \n",
       "min       0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "25%       0.000000     0.000000     0.000000     0.005869     0.000000   \n",
       "50%       0.007271     0.000000     0.000660     0.033716     0.002129   \n",
       "75%       0.036400     0.016172     0.008880     0.140159     0.010000   \n",
       "max       0.672066     0.946207     0.822799     1.471075     0.299035   \n",
       "\n",
       "            Notch1       Notch2       Notch3       Notch4         Mfng  \\\n",
       "count  1206.000000  1206.000000  1206.000000  1206.000000  1206.000000   \n",
       "mean      0.254916     0.270746     0.085560     0.048047     0.114217   \n",
       "std       0.327232     0.277195     0.250395     0.109029     0.172868   \n",
       "min       0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "25%       0.050362     0.058762     0.000263     0.003831     0.002875   \n",
       "50%       0.132765     0.166685     0.014431     0.014120     0.030390   \n",
       "75%       0.301811     0.406283     0.066928     0.036206     0.178138   \n",
       "max       1.567202     1.287853     1.770191     0.605964     0.853848   \n",
       "\n",
       "              Rfng         Lfng  \n",
       "count  1206.000000  1206.000000  \n",
       "mean      0.115788     0.165934  \n",
       "std       0.070568     0.264093  \n",
       "min       0.000000     0.000000  \n",
       "25%       0.067396     0.009333  \n",
       "50%       0.097068     0.040172  \n",
       "75%       0.151794     0.181560  \n",
       "max       0.346051     1.167906  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute the 99th quantile per gene\n",
    "gene_quantiles = np.quantile(df_pathway, q = 0.99, axis = 0)\n",
    "\n",
    "# Clip the values such that any counts above the 99th quantile take its value\n",
    "x_clipped = quantile_saturation(df_pathway, gene_quantiles)\n",
    "x_clipped.describe() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now scale expression values in this DataFrame of \"clipped\" counts to lie between 0 and 1, using a simple MinMaxScaler from scikitlearn that linearly scales counts to lie between the a specified minimum and maximum value per gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dll1</th>\n",
       "      <th>Dll3</th>\n",
       "      <th>Dll4</th>\n",
       "      <th>Jag1</th>\n",
       "      <th>Jag2</th>\n",
       "      <th>Notch1</th>\n",
       "      <th>Notch2</th>\n",
       "      <th>Notch3</th>\n",
       "      <th>Notch4</th>\n",
       "      <th>Mfng</th>\n",
       "      <th>Rfng</th>\n",
       "      <th>Lfng</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_E6.5_8.5_Chan</th>\n",
       "      <td>0.014819</td>\n",
       "      <td>0.136275</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.041043</td>\n",
       "      <td>0.016864</td>\n",
       "      <td>0.046589</td>\n",
       "      <td>0.062934</td>\n",
       "      <td>0.064582</td>\n",
       "      <td>0.114331</td>\n",
       "      <td>0.026468</td>\n",
       "      <td>0.236436</td>\n",
       "      <td>0.015711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_E6.5_8.5_Chan</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.068862</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005510</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001764</td>\n",
       "      <td>0.009885</td>\n",
       "      <td>0.008756</td>\n",
       "      <td>0.071154</td>\n",
       "      <td>0.034318</td>\n",
       "      <td>0.182304</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_E6.5_8.5_Chan</th>\n",
       "      <td>0.010974</td>\n",
       "      <td>0.073606</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036166</td>\n",
       "      <td>0.429603</td>\n",
       "      <td>0.062882</td>\n",
       "      <td>0.062029</td>\n",
       "      <td>0.063305</td>\n",
       "      <td>0.188277</td>\n",
       "      <td>0.025875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_E6.5_8.5_Chan</th>\n",
       "      <td>0.006131</td>\n",
       "      <td>0.088485</td>\n",
       "      <td>0.002994</td>\n",
       "      <td>0.008539</td>\n",
       "      <td>0.007506</td>\n",
       "      <td>0.032810</td>\n",
       "      <td>0.032471</td>\n",
       "      <td>0.059412</td>\n",
       "      <td>0.101778</td>\n",
       "      <td>0.073151</td>\n",
       "      <td>0.238092</td>\n",
       "      <td>0.027696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_E6.5_8.5_Chan</th>\n",
       "      <td>0.001608</td>\n",
       "      <td>0.057720</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001342</td>\n",
       "      <td>0.033267</td>\n",
       "      <td>0.015496</td>\n",
       "      <td>0.037971</td>\n",
       "      <td>0.001749</td>\n",
       "      <td>0.160801</td>\n",
       "      <td>0.007979</td>\n",
       "      <td>0.429918</td>\n",
       "      <td>0.001190</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Dll1      Dll3      Dll4      Jag1      Jag2    Notch1  \\\n",
       "1_E6.5_8.5_Chan  0.014819  0.136275  0.000000  0.041043  0.016864  0.046589   \n",
       "2_E6.5_8.5_Chan  0.000000  0.068862  0.000000  0.005510  0.000000  0.001764   \n",
       "3_E6.5_8.5_Chan  0.010974  0.073606  0.000000  0.040344  0.000000  0.036166   \n",
       "4_E6.5_8.5_Chan  0.006131  0.088485  0.002994  0.008539  0.007506  0.032810   \n",
       "5_E6.5_8.5_Chan  0.001608  0.057720  0.000000  0.001342  0.033267  0.015496   \n",
       "\n",
       "                   Notch2    Notch3    Notch4      Mfng      Rfng      Lfng  \n",
       "1_E6.5_8.5_Chan  0.062934  0.064582  0.114331  0.026468  0.236436  0.015711  \n",
       "2_E6.5_8.5_Chan  0.009885  0.008756  0.071154  0.034318  0.182304  0.000000  \n",
       "3_E6.5_8.5_Chan  0.429603  0.062882  0.062029  0.063305  0.188277  0.025875  \n",
       "4_E6.5_8.5_Chan  0.032471  0.059412  0.101778  0.073151  0.238092  0.027696  \n",
       "5_E6.5_8.5_Chan  0.037971  0.001749  0.160801  0.007979  0.429918  0.001190  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize the scaler\n",
    "scaler = MinMaxScaler()\n",
    "# Fit the scaler to the dataset to compute per-feature min. and max. values\n",
    "scaler.fit(x_clipped)\n",
    "\n",
    "# Apply the scaler to the dataset\n",
    "df_scale = pd.DataFrame(scaler.transform(x_clipped), columns = x_clipped.columns, index = x_clipped.index)\n",
    "\n",
    "df_scale.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notch Correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at gene correlation for clusters that have the BMP receptors active. We can try to measure this by first computing correlation only at clusters that have expression of 2 receptors greater than or equal to 0.2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1123, 12)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filt = pd.DataFrame(columns = df_scale.columns)\n",
    "\n",
    "for i in df_scale.index:\n",
    "    if (df_scale.loc[i][pathway_genes] >= 0.2).sum() > 0:\n",
    "        df_filt = df_filt.append(df_scale.loc[i])\n",
    "        \n",
    "df_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_corr = as.data.frame(cor(df_filt))\n",
    "colnames(M_corr) = colnames(df_filt)\n",
    "rownames(M_corr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/notch_corrplot_all.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_corr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notch Partial Correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n"
     ]
    }
   ],
   "source": [
    "%%R -i df_filt\n",
    "M_pcorr = as.data.frame(pcor(df_filt)$estimate)\n",
    "colnames(M_pcorr) = colnames(df_filt)\n",
    "rownames(M_pcorr) = colnames(df_filt)\n",
    "\n",
    "pdf(file = \"./alejo_metadata/figures/own_norm/notch_parcorrplot_all.pdf\", useDingbats=FALSE)\n",
    "corrplot(as.matrix(M_pcorr), type = \"upper\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "%R -o M_corr,M_pcorr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_corr = M_corr.where(np.triu(np.ones(M_corr.shape), k = 1).astype(np.bool))\n",
    "M_pcorr = M_pcorr.where(np.triu(np.ones(M_pcorr.shape), k = 1).astype(np.bool))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_max = np.zeros((2,2))\n",
    "\n",
    "min_max[0] = min(np.nanmin(M_corr.min(skipna = True)), np.nanmin(M_pcorr.min(skipna = True)))\n",
    "min_max[1] = max(np.nanmax(M_corr.max(skipna = True)), np.nanmax(M_pcorr.max(skipna = True)))\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for i in list(itertools.combinations(df_filt.columns, 2)):\n",
    "    x = pd.DataFrame()\n",
    "    df = df.append([[M_corr.loc[i[0]][i[1]], M_pcorr.loc[i[0]][i[1]], i]])\n",
    "    \n",
    "df.columns = [\"corr\", \"pcorr\", \"id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"aa7632dc-7449-479d-a730-74369d385035\" data-root-id=\"1905\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "    \n",
       "  var docs_json = {\"a391cfea-d02b-4c5e-8b6f-f6a47c529ab1\":{\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"1916\"}],\"center\":[{\"id\":\"1919\"},{\"id\":\"1923\"}],\"left\":[{\"id\":\"1920\"}],\"plot_height\":400,\"plot_width\":400,\"renderers\":[{\"id\":\"1935\"},{\"id\":\"1940\"}],\"title\":{\"id\":\"1906\"},\"toolbar\":{\"id\":\"1925\"},\"toolbar_location\":\"below\",\"x_range\":{\"id\":\"1908\"},\"x_scale\":{\"id\":\"1912\"},\"y_range\":{\"id\":\"1910\"},\"y_scale\":{\"id\":\"1914\"}},\"id\":\"1905\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1938\",\"type\":\"Line\"},{\"attributes\":{},\"id\":\"1910\",\"type\":\"DataRange1d\"},{\"attributes\":{},\"id\":\"1980\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1976\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{},\"id\":\"1921\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"1914\",\"type\":\"LinearScale\"},{\"attributes\":{},\"id\":\"1974\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{},\"id\":\"1908\",\"type\":\"DataRange1d\"},{\"attributes\":{\"data\":{\"corr\":{\"__ndarray__\":\"bJ7Pfgd93z+jC9NqQQjfP65YO4VmJ8g/TXvJkZ2L1T8NNPf5rInVP6ZSiFRHK8m/zxwx706clD9hZ+zD5YTfP2TfuEKVtc4/+w+fTnuJwD8e/97ifnHCPxmKupNkOra/i1ndk/Cmr79c4/tlYee3v+tEEY6epJa/EcHVIWvxx7/9r5C7j+6fv/s3lbdiA4q/Z7e1LQKDur8183++v8exvzXamPcZmaG/uMAoDEu2zz8fhAnV6GniP6Ig+jz1keI/35xfxPJZxb+BRoxJTGzAP2goB5GGaew/u8gU+P9a4j/QAlkGdPi8P9KH6nkLDMI/oR0IT+oZ4D8HNIGVQx7eP13oDMfAKbw/6sefhnRi3T9chQfPTs/RP3EJlv3dfZm/nOWH0MxDxj+8bFIoeR+9v+VH+BoT5d8/hEQgOcV8vr9EJ4xR+6+/P1wQ9P6JMOI/cB9qzeui0j+0j3jjUj7KP65FLSM7LqM/GWJIZJ2Dcz/VOTvLOI7PPxbHYr3N+eA/oLg5oLE+1T+XHAxINpmgPwAWlAMMMLI/WA4E6XOtuz+5XcqnDyLGvzouYCL007G/1Bxhw7dUqr+LtmLwJIGXP1oyfuYgBsY/b0jUQvmXqr+dTLnf4KDFPx0I40YRjLu/SJgB50ys4j8e6Ejia9HFPwSjmmyV2MM/jwNaIEDZt7+wgaBmL+DgP9ALrwviar6/\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[66]},\"id\":[[\"Dll1\",\"Dll3\"],[\"Dll1\",\"Dll4\"],[\"Dll1\",\"Jag1\"],[\"Dll1\",\"Jag2\"],[\"Dll1\",\"Notch1\"],[\"Dll1\",\"Notch2\"],[\"Dll1\",\"Notch3\"],[\"Dll1\",\"Notch4\"],[\"Dll1\",\"Mfng\"],[\"Dll1\",\"Rfng\"],[\"Dll1\",\"Lfng\"],[\"Dll3\",\"Dll4\"],[\"Dll3\",\"Jag1\"],[\"Dll3\",\"Jag2\"],[\"Dll3\",\"Notch1\"],[\"Dll3\",\"Notch2\"],[\"Dll3\",\"Notch3\"],[\"Dll3\",\"Notch4\"],[\"Dll3\",\"Mfng\"],[\"Dll3\",\"Rfng\"],[\"Dll3\",\"Lfng\"],[\"Dll4\",\"Jag1\"],[\"Dll4\",\"Jag2\"],[\"Dll4\",\"Notch1\"],[\"Dll4\",\"Notch2\"],[\"Dll4\",\"Notch3\"],[\"Dll4\",\"Notch4\"],[\"Dll4\",\"Mfng\"],[\"Dll4\",\"Rfng\"],[\"Dll4\",\"Lfng\"],[\"Jag1\",\"Jag2\"],[\"Jag1\",\"Notch1\"],[\"Jag1\",\"Notch2\"],[\"Jag1\",\"Notch3\"],[\"Jag1\",\"Notch4\"],[\"Jag1\",\"Mfng\"],[\"Jag1\",\"Rfng\"],[\"Jag1\",\"Lfng\"],[\"Jag2\",\"Notch1\"],[\"Jag2\",\"Notch2\"],[\"Jag2\",\"Notch3\"],[\"Jag2\",\"Notch4\"],[\"Jag2\",\"Mfng\"],[\"Jag2\",\"Rfng\"],[\"Jag2\",\"Lfng\"],[\"Notch1\",\"Notch2\"],[\"Notch1\",\"Notch3\"],[\"Notch1\",\"Notch4\"],[\"Notch1\",\"Mfng\"],[\"Notch1\",\"Rfng\"],[\"Notch1\",\"Lfng\"],[\"Notch2\",\"Notch3\"],[\"Notch2\",\"Notch4\"],[\"Notch2\",\"Mfng\"],[\"Notch2\",\"Rfng\"],[\"Notch2\",\"Lfng\"],[\"Notch3\",\"Notch4\"],[\"Notch3\",\"Mfng\"],[\"Notch3\",\"Rfng\"],[\"Notch3\",\"Lfng\"],[\"Notch4\",\"Mfng\"],[\"Notch4\",\"Rfng\"],[\"Notch4\",\"Lfng\"],[\"Mfng\",\"Rfng\"],[\"Mfng\",\"Lfng\"],[\"Rfng\",\"Lfng\"]],\"index\":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],\"pcorr\":{\"__ndarray__\":\"BYmXzcqb4z+KhOYyBVbQP0nK3Hr4mbs/9bRpJ0wipj+FWZAjIjqiP26Uchv4YZG//YGjCs6fv79ASt41niutP2BqBt5RjLW/3kQOcVlcwj+N7lweN/vGP3AZlX9o6dG/LfKD8SOytr+Q7Emxqpu1v3wXkctTXKY/oX6chNDSwr8W627p7OCvP+jQv9jujrs/v9J1mRhwpr+C7L+gOXHEv21rb1oWrLa/DnXfoyaMwr/ADm8Y3zrBPwzDUuDmcNA/LmqQ2ELErb9u7yhLH1qEP+C3ZeBVd+Y/AdNcpVoPwD/EnWBaOrG3vw2zjiaPo76/IsT2wVEc2j+nLlJ9S5vSPxfKOCNyw8A/DWFBFBAA2T9zaLoqQyeyP8azwka4SsK/TDOOVvBDSL+zqHIUPG2lv5LuPCMm/7o/C3+xujHUu78A8mKs0e/Hv7Deael447g/Ob8fsMkpkz9YT73hWWm+P2/A63cYL3i/kWMVG9Fzsz9IU11iWqGyP9YmiGlXEKe/HFYj8tbcsz8wrN0IYwa7v5YIUafpjH6/ZbpM9VPWrT+7QACbv0Cmvzs+BiHENJg/T/Q5e9QSob8WwMVy8SypP8knn9r3q7w/7JS6d10xr79YePOnAWi+P4ki+cFJAYK/bJfZv68H0T/UQ4xGuwDEP31+QJub4Ke//ymVm0bOwr+/AZpwkZfgP6p+6jNUl6W/\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[66]}},\"selected\":{\"id\":\"1978\"},\"selection_policy\":{\"id\":\"1979\"}},\"id\":\"1898\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1934\",\"type\":\"Scatter\"},{\"attributes\":{},\"id\":\"1979\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"callback\":null,\"tooltips\":[[\"Genes\",\"@id\"]]},\"id\":\"1899\",\"type\":\"HoverTool\"},{\"attributes\":{\"axis_label\":\"Correlation\",\"formatter\":{\"id\":\"1976\"},\"ticker\":{\"id\":\"1917\"}},\"id\":\"1916\",\"type\":\"LinearAxis\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"right_units\":\"screen\",\"top_units\":\"screen\"},\"id\":\"1924\",\"type\":\"BoxAnnotation\"},{\"attributes\":{\"data_source\":{\"id\":\"1898\"},\"glyph\":{\"id\":\"1933\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1934\"},\"selection_glyph\":null,\"view\":{\"id\":\"1936\"}},\"id\":\"1935\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"active_drag\":\"auto\",\"active_inspect\":\"auto\",\"active_multi\":null,\"active_scroll\":\"auto\",\"active_tap\":\"auto\",\"tools\":[{\"id\":\"1899\"},{\"id\":\"1900\"},{\"id\":\"1901\"},{\"id\":\"1902\"},{\"id\":\"1903\"},{\"id\":\"1904\"}]},\"id\":\"1925\",\"type\":\"Toolbar\"},{\"attributes\":{\"data\":{\"x\":{\"__ndarray__\":\"cBmVf2jp0b9oKAeRhmnsPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]},\"y\":{\"__ndarray__\":\"cBmVf2jp0b9oKAeRhmnsPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[2]}},\"selected\":{\"id\":\"1980\"},\"selection_policy\":{\"id\":\"1981\"}},\"id\":\"1937\",\"type\":\"ColumnDataSource\"},{\"attributes\":{},\"id\":\"1978\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"1900\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"1901\",\"type\":\"PanTool\"},{\"attributes\":{\"data_source\":{\"id\":\"1937\"},\"glyph\":{\"id\":\"1938\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"1939\"},\"selection_glyph\":null,\"view\":{\"id\":\"1941\"}},\"id\":\"1940\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"text\":\"Notch\"},\"id\":\"1906\",\"type\":\"Title\"},{\"attributes\":{\"fill_color\":{\"value\":\"#1f77b4\"},\"line_color\":{\"value\":\"#1f77b4\"},\"x\":{\"field\":\"corr\"},\"y\":{\"field\":\"pcorr\"}},\"id\":\"1933\",\"type\":\"Scatter\"},{\"attributes\":{\"source\":{\"id\":\"1898\"}},\"id\":\"1936\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"1903\",\"type\":\"ResetTool\"},{\"attributes\":{\"axis\":{\"id\":\"1920\"},\"dimension\":1,\"ticker\":null},\"id\":\"1923\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"1912\",\"type\":\"LinearScale\"},{\"attributes\":{},\"id\":\"1917\",\"type\":\"BasicTicker\"},{\"attributes\":{\"line_alpha\":0.1,\"x\":{\"field\":\"x\"},\"y\":{\"field\":\"y\"}},\"id\":\"1939\",\"type\":\"Line\"},{\"attributes\":{\"source\":{\"id\":\"1937\"}},\"id\":\"1941\",\"type\":\"CDSView\"},{\"attributes\":{\"axis\":{\"id\":\"1916\"},\"ticker\":null},\"id\":\"1919\",\"type\":\"Grid\"},{\"attributes\":{\"overlay\":{\"id\":\"1924\"}},\"id\":\"1902\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"1981\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"axis_label\":\"Partial Correlation\",\"formatter\":{\"id\":\"1974\"},\"ticker\":{\"id\":\"1921\"}},\"id\":\"1920\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"1904\",\"type\":\"SaveTool\"}],\"root_ids\":[\"1905\"]},\"title\":\"Bokeh Application\",\"version\":\"2.1.1\"}};\n",
       "  var render_items = [{\"docid\":\"a391cfea-d02b-4c5e-8b6f-f6a47c529ab1\",\"root_ids\":[\"1905\"],\"roots\":{\"1905\":\"aa7632dc-7449-479d-a730-74369d385035\"}}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "1905"
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'/Users/niveditakanrar/Documents/SURF/Code/alejo_metadata/figures/own_norm/notch_int_corr_pcorr_scatter.html'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source = ColumnDataSource(df)\n",
    "output_file(filename=figdir+\"notch_int_corr_pcorr_scatter.html\")\n",
    "\n",
    "hover = HoverTool(tooltips=[('Genes', '@id')])\n",
    "tools = [hover, WheelZoomTool(), PanTool(), BoxZoomTool(), ResetTool(), SaveTool()]\n",
    "\n",
    "p = figure(width=400, height=400,tools=tools,\n",
    "           toolbar_location=\"below\", \n",
    "           x_axis_label = \"Correlation\",\n",
    "           y_axis_label = \"Partial Correlation\",\n",
    "           title = \"Notch\")\n",
    "\n",
    "p.scatter(x = \"corr\", y = \"pcorr\", source = source, color = \"#1f77b4\");\n",
    "p.line(min_max[:,0], min_max[:,1], line_color = \"black\");\n",
    "\n",
    "show(p)\n",
    "save(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
