---
title: 'Example Pathway Analysis: TGF-β'
output:
  html_document:
    df_print: paged
---

Here, we implement our pipeline for analysis of gene set expression profiles using the TGF-β signaling pathway. 

We can begin with importing the necessary packages and setting up our directory names.

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/Users/niveditakanrar/Documents/Elowitz Lab/motifs")) 
```

```{r}
source("./scripts/analysis/imports.R")

## Change these variables
filename = "tgfb" # how name is appended to files for pathway
pathway_index = 2 # Tgfb pathway index

## Directories for the files
output_dir = paste("./scripts/analysis/outputs/", filename, "_analysis/", filename, sep = "")
fig_dir = "./scripts/figures/"
```

The integrated atlas is stored in the Seurat object `master_seurat` with cell states as observations and the intersection of genes as variable names. Metadata annotations are also stored here. This object contains the log-normalized counts from all datasets.

```{r}
master_seurat
```

We store GenBank gene lists in the `all_pathways` DataFrame. We can call on the stored TGF-β gene list below.

```{r}
pathway_name = "Bmp_Tgfb" # Tgfb pathway name
genesPathway(pathway_name)
```

## Identifying Pathway Expression Profiles
To identify the number of unique pathway expression profiles, we will compute the silhouette score on clusters defined using the K-means algorithm. For clustering of gene expression data, instead of considering the log-normalized counts in cell states, we MinMax normalize the counts to lie in the range of $[0, 1]$ to ensure that each gene has the same dynamic range. However, due to the noisy nature of count data, we compute the maximum value of 1 to correspond to the 95th percentile of a gene's counts. All steps in the pipeline below operate on the level of these MinMax normalized counts, computed internally in the functions from our `module.R` file.

We can specify the threshold parameters for the signaling pathway: the minimum normalized expression level and minimum number of genes for a pathway to be classified as "ON" in a cell state. For the GenBank pathways like TGF-beta, these are stored in the list `param_list`.

```{r}
min_genes_pathway = 2 # Tgfb min. number of genes expressed
min_expr_threshold = 0.2 # Tgfb minimum expression threshold for gene to be on
```

We can compute the optimal number of expression profiles using the silhouette score.

```{r}
## Computing the optimal silhouette score
silh_plt = silhouettePlot(which_pathway = pathway_name, 
                          min_ON = min_genes_pathway, 
                          min_expr = min_expr_threshold, 
                          n_bootstraps = 100)

# Plot the silhouette score plot along with the Z-score plot
g <- silhouette_zscore(silh_result = silh_plt,
                       pathway_name = pathway_name,
                       x_offset = 6,
                       max.y = 0.43,
                       min.y =0.17 # adjust axis parameters 
)
g

optimal_k_pathway = 30 # optimal pathway components, computed from z-score
```

Now that we have extracted the profiles, we can look at their diversity. Downstream analysis of these profiles requires us to run the function `fullControlPathway` to extract cell state diversity.

```{r}
## Running fullControlPathway
control_res  <- fullControlPathway(this_pathway = pathway_name,
                                  k_pathway = optimal_k_pathway , 
                                  filter_pathway = 'both', # adult + devel 
                                  this_seurat = master_seurat, # seurat object
                                  null_list = hvg_genes, #list of highly variable genes 
                                  n_samples = 100, filter_manual = T,
                                  min_expr_gene = min_expr_threshold, 
                                  min_genes_ON = min_genes_pathway, # default > 2 genes ON 
                                  n_pcs_global = 100, # how many PCs to use
                                  embedding_matrix = pca_proj # PCA embedding for diversity 
                                  )
```

We can now plot various summary plots for profile diversity.

```{r}
# ECDF plots
ecdf_plot = ecdf_diversity(control_res)
ecdf_plot

# Diversity rank plots
rank_plot <- rank_diversity(
                            which_pathway = pathway_name, 
                            k_motifs = optimal_k_pathway, 
                            min_expression = min_expr_threshold, 
                            min_genes_pathway = min_genes_pathway,
                            embedding_matrix = pca_proj
                            )
rank_plot
```

We can also visualize the profiles directly on the cell state dendrogram.

```{r}
# Plot global dendrogram with pathway states
glob_dendr <- globalDendrogramPlot2(
                                    control_res = control_res, 
                                    seurat_obj = master_seurat, 
                                    use_pca = T, 
                                    npcs = 1:20,
                                    clust_method = 'ward.D2',
                                    dist_metric ='cosine',
                                    save_pdf = F
)
```

We can obtain the motif profiles, classified as profiles whose mean pairwise distances lie above the 90th percentile of distances.

```{r}
## Plot the most diverse profiles, aka the motifs
# 1. Select all profiles 
diverse_df <- control_res$profiles
# 2. tidy data.frame to average gene expression within a pathway profile 
diverse_df %>% pivot_longer(cols = genesPathway(pathway_name), 
                            names_to = 'gene', 
                            values_to = 'expression') %>% 
           select(cell_id, cell_ontology_class, Tissue, Cell_class, gene, expression, class_label, rank, diversity, n) -> diverse_tidy 

diverse_tidy %>% group_by(class_label,gene) %>% summarise(mean_expr = mean(expression), 
            rank = mean(rank),diversity = mean(diversity), 
            cell_types = mean(n)) -> diverse_tidy

# 3. wide format 
diverse_tidy %>% pivot_wider(id_cols = c(class_label,rank, diversity, cell_types), names_from=gene,values_from=mean_expr) %>% tibble::column_to_rownames('class_label')-> diverse_mat

# 4. We do the filtering here either for motifs or for non-diverse profiles 
control_res$diversity %>% 
  dplyr::filter(type=='transcriptome') %>% # choose the null model 
  pull(d) %>% quantile(0.90) -> divers_quantile

diverse_mat %>% dplyr::filter(diversity>divers_quantile) -> diverse_mat

motif_heatmap <- superheat(diverse_mat[,genesPathway(pathway_name)],
                           pretty.order.rows = T,
                           heat.pal = black_pal(10),
                           bottom.label.text.angle = 90, 
                           yr = sqrt(diverse_mat$cell_types),
                           yr.plot.type='bar',
                           yr.axis.name = "N cell types",
                           row.title = "Pathway motifs",
                           column.title = "Pathway genes",
                           bottom.label.size = 0.3,
                           grid.hline.col = "white",
                           grid.hline.size = 2, 
                           
)
```

We can also visualize the distribution of these motifs across tissues and organs.

```{r}
# 6. Tissue and organ distributions of selected profiles
diverse_df %>% dplyr::filter(class_label %in% row.names(diverse_mat) ) %>% 
  select(Tissue, class_label ) %>% dplyr::filter(!Tissue %in% c('mat','endoderm')) %>% group_by(class_label,Tissue)%>% 
  count  %>% 
  pivot_wider(id_cols=class_label,names_from = Tissue,
              values_from= n,values_fill = 0) -> tissue_distribution

tissue_pal<-colorRampPalette(brewer.pal(n = 9, name = 'PuRd'))
x = tissue_distribution %>% ungroup %>% select(-class_label) %>% as.matrix() 
row.names(x) <- tissue_distribution$class_label
# make tissue names pretty 
colnames(x)<-str_replace(string = colnames(x),pattern ="_",replacement = ' ') %>% firstup() 

x <- x[,sort(colnames(x))]

pheatmap(sqrt(x), treeheight_row = 20,treeheight_col = 20,
         clustering_method = 'ward.D2',col = tissue_pal(100),
         cluster_rows = F,
         cluster_cols = F,
         fontsize =12,angle_col = 45,
         height =4, width = 6) # keep this size to make it similar to the motif heatmap
```

